Index: etc/rc.subr
===================================================================
--- etc/rc.subr	(revision 227454)
+++ etc/rc.subr	(working copy)
@@ -700,9 +700,7 @@ run_rc_command()
 				if [ -n "${rc_quiet}" ]; then
 					return 0
 				fi
-				echo -n "Cannot '${rc_arg}' $name. Set ${rcvar} to "
-				echo -n "YES in /etc/rc.conf or use 'one${rc_arg}' "
-				echo "instead of '${rc_arg}'."
+				echo "Will not '${rc_arg}' $name because ${rcvar} is NO."
 				return 0
 			fi
 		fi
