Index: sys/dev/mfi/mfi.c
===================================================================
--- sys/dev/mfi/mfi.c	(revision 226546)
+++ sys/dev/mfi/mfi.c	(working copy)
@@ -928,6 +928,12 @@ mfi_intr(void *arg)
 	if (sc->mfi_check_clear_intr(sc))
 		return;
 
+	/*
+	 * Do a dummy read to flush the interrupt ACK that we just performed,
+	 * ensuring that everything is really, truly consistent.
+	 */
+	(void)sc->mfi_read_fw_status(sc);
+
 	pi = sc->mfi_comms->hw_pi;
 	ci = sc->mfi_comms->hw_ci;
 	mtx_lock(&sc->mfi_io_lock);
