Index: rescue/rescue/Makefile
===================================================================
--- FreeBSD/src/rescue/rescue/Makefile	2010-06-13 20:09:06.000000000 -0600
+++ FreeBSD/src/rescue/rescue/Makefile	2010-09-08 11:37:20.000000000 -0600
@@ -200,6 +200,7 @@
 CRUNCH_SRCDIRS+= usr.bin
 
 CRUNCH_PROGS_usr.bin= head mt sed tail tee
+CRUNCH_PROGS_usr.bin+= awk cmp cut dialog dirname find grep mktemp tr uname
 
 CRUNCH_PROGS_usr.bin+= gzip
 CRUNCH_ALIAS_gzip= gunzip gzcat zcat
@@ -224,12 +225,16 @@
 CRUNCH_PROGS_usr.bin+= id
 CRUNCH_ALIAS_id= groups whoami
 
+CRUNCH_SRCDIR_dialog= ${.CURDIR}/../../gnu/usr.bin/dialog
+CRUNCH_LIBS+= -ldialog -lncurses
+
 ##################################################################
 # Programs from stock /usr/sbin
 #
 CRUNCH_SRCDIRS+= usr.sbin
 
 CRUNCH_PROGS_usr.sbin= chroot
+CRUNCH_PROGS_usr.sbin+= vidcontrol
 
 CRUNCH_PROGS_usr.sbin+= chown
 CRUNCH_ALIAS_chown= chgrp
