diff --git sys/net/if_lagg.c sys/net/if_lagg.c
index 59237bc..dea45d9 100644
--- sys/net/if_lagg.c
+++ sys/net/if_lagg.c
@@ -485,7 +485,9 @@ lagg_port_setlladdr(void *arg, int pending)
 		ifp = llq->llq_ifp;
 
 		/* Set the link layer address */
+		CURVNET_SET(ifp->if_vnet);
 		error = if_setlladdr(ifp, llq->llq_lladdr, ETHER_ADDR_LEN);
+		CURVNET_RESTORE();
 		if (error)
 			printf("%s: setlladdr failed on %s\n", __func__,
 			    ifp->if_xname);
