diff --git sys/net/if_lagg.c sys/net/if_lagg.c
index bc5618f..5dde7b2 100644
--- sys/net/if_lagg.c
+++ sys/net/if_lagg.c
@@ -1048,6 +1048,23 @@ lagg_ioctl(struct ifnet *ifp, u_long cmd, caddr_t data)
 		break;
 
 	case SIOCSIFCAP:
+		LAGG_WLOCK(sc);
+		/* Set caps on ports too */
+		SLIST_FOREACH(lp, &sc->sc_ports, lp_entries) {
+			if (lp->lp_ioctl == NULL) {
+				error = EINVAL;
+				break;
+			}
+			error = (*lp->lp_ioctl)(lp->lp_ifp, cmd, data);
+			if (error) {
+				break;
+			}
+		}
+		if (error == 0)
+			lagg_capabilities(sc);
+		LAGG_WUNLOCK(sc);
+		break;
+
 	case SIOCSIFMTU:
 		/* Do not allow the MTU or caps to be directly changed */
 		error = EINVAL;
