Index: sys/netinet/ip_carp.c
===================================================================
--- sys/netinet/ip_carp.c	(revision 244186)
+++ sys/netinet/ip_carp.c	(working copy)
@@ -871,22 +871,27 @@
 		if (sc->sc_carpdev == NULL)
 			continue;
 		CARP_SCLOCK(sc);
+		CURVNET_SET(sc->sc_carpdev->if_vnet);
 		if ((SC2IFP(sc)->if_flags & IFF_UP) &&
 		    (SC2IFP(sc)->if_drv_flags & IFF_DRV_RUNNING) &&
 		     sc->sc_state == MASTER)
 			carp_send_ad_locked(sc);
+		CURVNET_RESTORE();
 		CARP_SCUNLOCK(sc);
 	}
 	mtx_unlock(&carp_mtx);
 }
 
+/* Send a periodic advertisement, executed in callout context. */
 static void
 carp_send_ad(void *v)
 {
 	struct carp_softc *sc = v;
 
 	CARP_SCLOCK(sc);
+	CURVNET_SET(sc->sc_carpdev->if_vnet);
 	carp_send_ad_locked(sc);
+	CURVNET_RESTORE();
 	CARP_SCUNLOCK(sc);
 }
 
@@ -1327,13 +1332,16 @@
 	return (NULL);
 }
 
+/* Master down timeout event, executed in callout context. */
 static void
 carp_master_down(void *v)
 {
 	struct carp_softc *sc = v;
 
 	CARP_SCLOCK(sc);
+	CURVNET_SET(sc->sc_carpdev->if_vnet);
 	carp_master_down_locked(sc);
+	CURVNET_RESTORE();
 	CARP_SCUNLOCK(sc);
 }
 
