Index: rescue/rescue/Makefile
===================================================================
--- FreeBSD/src/rescue/rescue/Makefile	2010-06-13 20:09:06.000000000 -0600
+++ FreeBSD/src/rescue/rescue/Makefile	2010-09-08 11:37:20.000000000 -0600
@@ -72,6 +72,7 @@
 CRUNCH_PROGS_bin= cat chflags chio chmod cp date dd df echo 	\
 	 ed expr getfacl hostname kenv kill ln ls mkdir mv	\
 	 pkill ps pwd realpath rm rmdir setfacl sh stty sync test
+CRUNCH_PROGS_bin+=sleep
 CRUNCH_LIBS+= -lcrypt -ledit -lkvm -ll -ltermcap -lutil
 
 # Additional options for specific programs
@@ -200,6 +201,7 @@
 CRUNCH_SRCDIRS+= usr.bin
 
 CRUNCH_PROGS_usr.bin= head mt sed tail tee
+CRUNCH_PROGS_usr.bin+= awk cmp cut dialog dirname find grep mktemp sort tr uname
 
 CRUNCH_PROGS_usr.bin+= gzip
 CRUNCH_ALIAS_gzip= gunzip gzcat zcat
@@ -224,12 +226,17 @@
 CRUNCH_PROGS_usr.bin+= id
 CRUNCH_ALIAS_id= groups whoami
 
+CRUNCH_SRCDIR_dialog= ${.CURDIR}/../../gnu/usr.bin/dialog
+CRUNCH_LIBS+= -ldialog -lncurses
+CRUNCH_SRCDIR_sort= ${.CURDIR}/../../gnu/usr.bin/sort
+
 ##################################################################
 # Programs from stock /usr/sbin
 #
 CRUNCH_SRCDIRS+= usr.sbin
 
 CRUNCH_PROGS_usr.sbin= chroot
+CRUNCH_PROGS_usr.sbin+= vidcontrol
 
 CRUNCH_PROGS_usr.sbin+= chown
 CRUNCH_ALIAS_chown= chgrp
