--- sys/cam/cam_xpt.c~	2012-03-15 08:45:24.891950142 -0700
+++ sys/cam/cam_xpt.c	2012-03-15 08:45:47.181884939 -0700
@@ -751,12 +751,14 @@
 			 * passthrough driver, but doesn't have one in his
 			 * kernel.
 			 */
+#if 0
 			if (base_periph_found == 1) {
 				printf("xptioctl: pass driver is not in the "
 				       "kernel\n");
 				printf("xptioctl: put \"device pass\" in "
 				       "your kernel config file\n");
 			}
+#endif
 		}
 		mtx_unlock(&xsoftc.xpt_topo_lock);
 		break;
