Index: rescue/rescue/Makefile
===================================================================
--- FreeBSD/src/rescue/rescue/Makefile	2010-06-13 20:09:06.000000000 -0600
+++ FreeBSD/src/rescue/rescue/Makefile	2010-09-08 11:37:20.000000000 -0600
@@ -199,7 +199,7 @@
 #
 CRUNCH_SRCDIRS+= usr.bin
 
-CRUNCH_PROGS_usr.bin= head mt sed tail tee
+CRUNCH_PROGS_usr.bin= awk dialog grep head mt sed tail tee
 
 CRUNCH_PROGS_usr.bin+= gzip
 CRUNCH_ALIAS_gzip= gunzip gzcat zcat
@@ -224,6 +224,11 @@
 CRUNCH_PROGS_usr.bin+= id
 CRUNCH_ALIAS_id= groups whoami
 
+CRUNCH_SRCDIR_dialog= ${.CURDIR}/../../gnu/usr.bin/dialog
+CRUNCH_LIBS+= -ldialog -lncurses
+CRUNCH_SRCDIR_grep= ${.CURDIR}/../../gnu/usr.bin/grep
+CRUNCH_LIBS+= -lgnuregex
+
 ##################################################################
 # Programs from stock /usr/sbin
 #
