--- tools/tools/nanobsd/nanobsd.sh~	2011-09-08 17:52:28.000000000 -0700
+++ tools/tools/nanobsd/nanobsd.sh	2011-09-08 17:53:04.000000000 -0700
@@ -366,8 +366,8 @@
 	dev=$1
 	dir=$2
 	mnt=$3
-	test -z $2 && dir=/var/empty
-	test -d $d || dir=/var/empty
+	test -z $2 && dir=${NANO_WORLDDIR}/var/empty
+	test -d $d || dir=${NANO_WORLDDIR}/var/empty
 	echo "Creating ${dev} with ${dir} (mounting on ${mnt})"
 	newfs ${NANO_NEWFS} ${dev}
 	mount ${dev} ${mnt}
