Index: sys/cddl/compat/opensolaris/sys/sid.h
===================================================================
--- sys/cddl/compat/opensolaris/sys/sid.h	(revision 240258)
+++ sys/cddl/compat/opensolaris/sys/sid.h	(revision 240259)
@@ -30,7 +30,8 @@
 #define	_OPENSOLARIS_SYS_SID_H_
 
 typedef struct ksiddomain {
-	char	kd_name[1];	/* Domain part of SID */
+	char	*kd_name;	/* Domain part of SID */
+	uint_t	kd_len;
 } ksiddomain_t;
 typedef void	ksid_t;
 
@@ -38,8 +39,12 @@ static __inline ksiddomain_t *
 ksid_lookupdomain(const char *domain)
 {
 	ksiddomain_t *kd;
+	size_t len;
 
-	kd = kmem_alloc(sizeof(*kd) + strlen(domain), KM_SLEEP);
+	len = strlen(domain) + 1;
+	kd = kmem_alloc(sizeof(*kd), KM_SLEEP);
+	kd->kd_len = (uint_t)len;
+	kd->kd_name = kmem_alloc(len, KM_SLEEP);
 	strcpy(kd->kd_name, domain);
 	return (kd);
 }
@@ -48,6 +53,7 @@ static __inline void
 ksiddomain_rele(ksiddomain_t *kd)
 {
 
+	kmem_free(kd->kd_name, kd->kd_len);
 	kmem_free(kd, sizeof(*kd));
 }
 
Index: sys/cddl
===================================================================
--- sys/cddl	(revision 240258)
+++ sys/cddl	(revision 240259)

Property changes on: sys/cddl
___________________________________________________________________
Modified: svn:mergeinfo
   Merged /head/sys/cddl:r240162
Index: sys
===================================================================
--- sys	(revision 240258)
+++ sys	(revision 240259)

Property changes on: sys
___________________________________________________________________
Modified: svn:mergeinfo
   Merged /head/sys:r240162
