--- sysutils/testdisk/Makefile.orig	2013-04-02 15:25:42.662231392 -0600
+++ sysutils/testdisk/Makefile	2013-04-02 15:32:33.844277434 -0600
@@ -1,11 +1,12 @@
 # $FreeBSD$
 
 PORTNAME=	testdisk
-PORTVERSION=	6.13
+PORTVERSION=	6.14
 PORTREVISION=	1
 CATEGORIES=	sysutils
 MASTER_SITES=	http://www.cgsecurity.org/
-DISTFILES=	${PORTNAME}-${PORTVERSION}${EXTRACT_SUFX}
+DISTVERSIONSUFFIX= -WIP
+DISTFILES= 	${PORTNAME}-${PORTVERSION}-WIP${EXTRACT_SUFX}
 
 MAINTAINER=	flz@FreeBSD.org
 COMMENT=	Tool to check and undelete partitions
