--- ports-mgmt/pbi-manager/Makefile	2012-05-11 14:04:06.000000000 -0700
+++ ports-mgmt/pbi-manager/Makefile	2012-05-23 11:36:41.044786953 -0700
@@ -13,7 +13,9 @@
 MAINTAINER=	kmoore@FreeBSD.org
 COMMENT=	Push Button Installer Manager
 
+.if !defined(WITHOUT_X11)
 RUN_DEPENDS=	xdg-desktop-icon:${PORTSDIR}/devel/xdg-utils
+.endif
 
 USE_BZIP2=	yes
 NO_BUILD=	yes
--- /dev/null	1969-12-31 16:00:00.000000000 -0800
+++ ports-mgmt/pbi-manager/files/patch-pbi-manager	2012-05-23 11:39:34.914788481 -0700
@@ -0,0 +1,52 @@
+--- pbi-manager	2012-05-23 11:37:39.424788861 -0700
++++ pbi-manager	2012-05-23 11:25:56.984786225 -0700
+@@ -5551,6 +5551,7 @@
+ 		_flags="${_flags} --tmpfs"
+ 	fi
+ 
++        unset PBI_PROGVERSION
+ 	get_pbi_progversion
+ 
+ 	#echo "Starting build of ${_mp} - ${_pv}"
+@@ -6345,7 +6346,7 @@
+ mk_chroot_file() {
+ 
+ 	# Check if on PC-BSD and we can instead fetch fbsd-release.txz
+-	if [ -e "$PCBSD_ETCCONF" ]; then
++	if [ -e "$PCBSD_ETCCONF" -a -z "$FORCE_FBSD_ONLY" ]; then
+ 
+ 	  cd "$PBI_APPDIR"
+ 
+@@ -6412,6 +6413,16 @@
+ 	
+ 		# Figure out which version of FreeBSD to checkout
+ 		case $_osRel in
++			*ALPHA)
++				local _osMajor=${_osRel%%.*}
++				local svnUrl="svn://svn.freebsd.org/base/releng/${_osMajor}.0"
++				echo "Using $svnUrl for sources"
++				;;
++			*BETA)
++				local _osMajor=${_osRel%%.*}
++				local svnUrl="svn://svn.freebsd.org/base/releng/${_osMajor}.0"
++				echo "Using $svnUrl for sources"
++				;;
+ 			*RELEASE)
+ 				local _osMajor=${_osRel%%.*}
+ 				local svnUrl="svn://svn.freebsd.org/base/releng/${_osMajor}.0"
+@@ -6439,6 +6450,7 @@
+                 else
+                         svn co ${svnUrl} ${PBI_BUILDSRC} >>${PBI_BUILDLOG} 2>>${PBI_BUILDLOG}
+                 fi
++		touch ${PBI_BUILDSRC}/Makefile
+ 
+                 if [ "$?" != "0" ] ; then
+                         if [ "${PBI_DELETE_BUILD}" = "1" ] ; then
+@@ -6450,7 +6462,6 @@
+ 	fi # End of subversion checkout
+ 
+ 	echo "Running buildworld / installworld"
+-	touch ${PBI_BUILDSRC}/Makefile
+ 	cd ${PBI_BUILDSRC}
+         make ${PBI_BUILDFLAGS} buildworld >>${PBI_BUILDLOG} 2>>${PBI_BUILDLOG}
+         if [ "$?" != "0" ] ; then
