Index: net/netatalk/Makefile
===================================================================
RCS file: /home/ncvs/ports/net/netatalk/Makefile,v
retrieving revision 1.96
diff -u -r1.96 Makefile
--- net/netatalk/Makefile	3 Jul 2011 11:40:45 -0000	1.96
+++ net/netatalk/Makefile	10 Oct 2011 06:42:50 -0000
@@ -2,12 +2,11 @@
 # Date created:         23 Jul 1997
 # Whom:                 stb
 #
-# $FreeBSD: ports/net/netatalk/Makefile,v 1.96 2011/07/03 11:40:45 swills Exp $
+# $FreeBSD: ports/net/netatalk/Makefile,v 1.103 2011/09/10 18:40:45 marcus Exp $
 #
 
 PORTNAME=	netatalk
-PORTVERSION=	2.1.5
-PORTREVISION=	2
+PORTVERSION=	2.2.1
 PORTEPOCH=	1
 CATEGORIES=	net print
 MASTER_SITES=	SF
@@ -24,6 +23,7 @@
 GNU_CONFIGURE=	yes
 USE_GMAKE=	yes
 USE_PERL5=	yes
+WANT_GNOME=	yes
 USE_RC_SUBR=	netatalk
 
 CONFIGURE_ARGS+=	--with-tcp-wrappers \
@@ -37,24 +37,20 @@
 		SRVLOC		"Enable Service Location Protocol support" off \
 		PAM		"Enable PAM support" off \
 		TIMELORD	"Enable Timelord network time service" off \
-		KRB5		"Enable Kerberos V UAM" off
+		KRB5		"Enable Kerberos V UAM" off \
+		ZEROCONF	"Enable Zeroconf (Bonjour) support" on
 
-FILES=		AppleVolumes.default AppleVolumes.system afpd.conf \
-		atalkd.conf papd.conf netatalk.conf
+FILES=		AppleVolumes.default AppleVolumes.system afpd.conf netatalk.conf
 LINKS=		unbin unhex unsingle hqx2bin single2bin macbinary \
 		binheader nadheader
-MAN1=		achfile.1 ad.1 aecho.1 afile.1 afppasswd.1 apple_cp.1 \
-		apple_dump.1 apple_mv.1 apple_rm.1 asip-status.pl.1 dbd.1 \
-		getzones.1 hqx2bin.1 macbinary.1 megatron.1 nbp.1 nbplkup.1 \
-		nbprgstr.1 nbpunrgstr.1 netatalk-config.1 pap.1 papstatus.1 \
-		psorder.1 single2bin.1 unbin.1 unhex.1 uniconv.1 unsingle.1
-MAN3=		atalk_aton.3 nbp_name.3
-MAN4=		atalk.4
+MAN1=		ad.1 afpldaptest.1 afppasswd.1 apple_dump.1 asip-status.pl.1 \
+		dbd.1 hqx2bin.1 macbinary.1 macusers.1 megatron.1 \
+		netatalk-config.1 single2bin.1 unbin.1 unhex.1 uniconv.1 \
+		unsingle.1
 MAN5=		AppleVolumes.5 AppleVolumes.default.5 \
 		AppleVolumes.system.5 afp_ldap.conf.5 afp_signature.conf.5 \
-		afpd.conf.5 atalkd.conf.5 netatalk.conf.5 papd.conf.5
-MAN8=		afp_acls.8 afpd.8 atalkd.8 cnid_dbd.8 cnid_metad.8 papd.8 \
-		papstatus.8 psf.8 timelord.8
+		afp_voluuid.conf.5 afpd.conf.5 netatalk.conf.5
+MAN8=		afpd.8 cnid_dbd.8 cnid_metad.8 timelord.8
 
 CONFLICTS=	bigloo-2.* cap-6.0.* tct-1.* netatalk-1* yudit-[0-9]*
 
@@ -68,9 +64,17 @@
 .endif
 
 .if defined(WITH_APPLETALK)
-CONFIGURE_ARGS+=	--with-ddp
+CONFIGURE_ARGS+=	--enable-ddp
+PLIST_SUB+=		APPLETALK=""
+MAN1+=			aecho.1 getzones.1 nbp.1 nbplkup.1 nbprgstr.1 \
+			nbpunrgstr.1 pap.1 papstatus.1 psorder.1
+MAN3+=			atalk_aton.3 nbp_name.3
+MAN4+=			atalk.4
+MAN5+=			atalkd.conf.5 papd.conf.5
+MAN8+=			atalkd.8 papd.8 papstatus.8 psf.8
+FILES+=			atalkd.conf papd.conf
 .else
-CONFIGURE_ARGS+=	--without-ddp
+PLIST_SUB+=		APPLETALK="@comment "
 .endif
 
 .if defined(WITH_PAM)
@@ -96,9 +100,26 @@
 PLIST_SUB+=		TIMELORD="@comment "
 .endif
 
+.if defined (WITH_ZEROCONF)
+CONFIGURE_ARGS+=	--enable-zeroconf=${LOCALBASE}
+CFLAGS+=		-I${LOCALBASE}/include -L${LOCALBASE}/lib
+LIB_DEPENDS+=		avahi-client.3:${PORTSDIR}/net/avahi-app
+USE_GNOME+=		pkgconfig
+SUB_LIST+=		ZEROCONF="avahi_daemon"
+.else
+SUB_LIST+=		ZEROCONF=""
+.endif
+
+.if ${OSVERSION} < 800031
+PLIST_SUB+=	ATFUNCS="@comment "
+.else
+PLIST_SUB+=	ATFUNCS=""
+.endif
+
 post-patch:
 	@${REINPLACE_CMD} -e 's|%%DB_NAME%%|${BDB_INCLUDE_DIR:T}| ; \
-	    	s|%%DB_LIB%%|-l${BDB_LIB_NAME}|g' \
+	    	s|%%DB_LIB%%|-l${BDB_LIB_NAME}|g ; \
+		s|%%LOCALBASE%%|${LOCALBASE}|g' \
 		${WRKSRC}/configure
 
 post-install:
Index: net/netatalk/distinfo
===================================================================
RCS file: /home/ncvs/ports/net/netatalk/distinfo,v
retrieving revision 1.30
diff -u -r1.30 distinfo
--- net/netatalk/distinfo	29 Dec 2010 04:36:17 -0000	1.30
+++ net/netatalk/distinfo	10 Oct 2011 06:42:51 -0000
@@ -1,2 +1,2 @@
-SHA256 (netatalk-2.1.5.tar.bz2) = 11fcce36cc5179de60c5c0b10032ff9e042ed8b8c6e0b99d2d7200c8d0749038
-SIZE (netatalk-2.1.5.tar.bz2) = 1125946
+SHA256 (netatalk-2.2.1.tar.bz2) = 3ab81c6335f8c33fd01ae599459a26c8cfe3b975009ce73640cae823ddc78bbf
+SIZE (netatalk-2.2.1.tar.bz2) = 1227602
Index: net/netatalk/pkg-plist
===================================================================
RCS file: /home/ncvs/ports/net/netatalk/pkg-plist,v
retrieving revision 1.31
diff -u -r1.31 pkg-plist
--- net/netatalk/pkg-plist	2 Jun 2010 14:46:42 -0000	1.31
+++ net/netatalk/pkg-plist	10 Oct 2011 06:42:51 -0000
@@ -1,20 +1,15 @@
-bin/achfile
-bin/ad
+%%ATFUNCS%%bin/ad
 bin/add_netatalk_printer
 bin/adv1tov2
-bin/aecho
-bin/afile
+%%APPLETALK%%bin/aecho
+bin/afpldaptest
 bin/afppasswd
-bin/apple_cp
 bin/apple_dump
-bin/apple_mv
-bin/apple_rm
 bin/binheader
 bin/asip-status.pl
 bin/cnid2_create
 bin/dbd
-bin/getzones
-bin/logger_test
+%%APPLETALK%%bin/getzones
 bin/macusers
 bin/megatron
 @exec cd %B && ln -s megatron hqx2bin
@@ -31,14 +26,13 @@
 @unexec rm -f %B/unbin
 @unexec rm -f %B/unhex
 @unexec rm -f %B/unsingle
-bin/nbplkup
-bin/nbprgstr
-bin/nbpunrgstr
-bin/netacnv
+%%APPLETALK%%bin/nbplkup
+%%APPLETALK%%bin/nbprgstr
+%%APPLETALK%%bin/nbpunrgstr
 bin/netatalk-config
-bin/pap
-bin/papstatus
-bin/psorder
+%%APPLETALK%%bin/pap
+%%APPLETALK%%bin/papstatus
+%%APPLETALK%%bin/psorder
 bin/showppd
 bin/uniconv
 @unexec if cmp -s %D/etc/AppleVolumes.default %D/etc/AppleVolumes.default.dist; then rm -f %D/etc/AppleVolumes.default; fi
@@ -50,15 +44,18 @@
 @unexec if cmp -s %D/etc/afpd.conf %D/etc/afpd.conf.dist; then rm -f %D/etc/afpd.conf; fi
 etc/afpd.conf.dist
 @exec [ ! -f %B/afpd.conf ] && cp %B/%f %B/afpd.conf
-@unexec if cmp -s %D/etc/atalkd.conf %D/etc/atalkd.conf.dist; then rm -f %D/etc/atalkd.conf; fi
-etc/atalkd.conf.dist
-@exec [ ! -f %B/atalkd.conf ] && cp %B/%f %B/atalkd.conf
+@unexec if cmp -s %D/etc/afp_ldap.conf %D/etc/afp_ldap.conf.dist; then rm -f %D/etc/afp_ldap.conf; fi
+etc/afp_ldap.conf.dist
+@exec [ ! -f %B/afp_ldap.conf ] && cp %B/%f %B/afp_ldap.conf
+%%APPLETALK%%@unexec if cmp -s %D/etc/atalkd.conf %D/etc/atalkd.conf.dist; then rm -f %D/etc/atalkd.conf; fi
+%%APPLETALK%%etc/atalkd.conf.dist
+%%APPLETALK%%@exec [ ! -f %B/atalkd.conf ] && cp %B/%f %B/atalkd.conf
 @unexec if cmp -s %D/etc/netatalk.conf %D/etc/netatalk.conf.dist; then rm -f %D/etc/netatalk.conf; fi
 etc/netatalk.conf.dist
 @exec [ ! -f %B/netatalk.conf ] && cp %B/%f %B/netatalk.conf
 @unexec if cmp -s %D/etc/papd.conf %D/etc/papd.conf.dist; then rm -f %D/etc/papd.conf; fi
-etc/papd.conf.dist
-@exec [ ! -f %B/papd.conf ] && cp %B/%f %B/papd.conf
+%%APPLETALK%%etc/papd.conf.dist
+%%APPLETALK%%@exec [ ! -f %B/papd.conf ] && cp %B/%f %B/papd.conf
 libexec/netatalk-uams/uams_clrtxt.so
 libexec/netatalk-uams/uams_dhx.so
 libexec/netatalk-uams/uams_dhx2.so
@@ -108,6 +105,7 @@
 include/atalk/logger.h
 include/atalk/nbp.h
 include/atalk/netddp.h
+include/atalk/queue.h
 include/atalk/pap.h
 include/atalk/paths.h
 include/atalk/rtmp.h
@@ -146,7 +144,7 @@
 libexec/ofwmpap
 libexec/ofwpap
 libexec/psa
-libexec/psf
+%%APPLETALK%%libexec/psf
 libexec/tfmpap
 libexec/tfmpaprev
 libexec/tfpap
@@ -156,10 +154,10 @@
 libexec/tfwpap
 libexec/tfwpaprev
 sbin/afpd
-sbin/atalkd
+%%APPLETALK%%sbin/atalkd
 sbin/cnid_dbd
 sbin/cnid_metad
-sbin/papd
+%%APPLETALK%%sbin/papd
 %%TIMELORD%%sbin/timelord
 share/aclocal/netatalk.m4
 %%DATADIR%%/pagecount.ps
Index: net/netatalk/files/netatalk.in
===================================================================
RCS file: /home/ncvs/ports/net/netatalk/files/netatalk.in,v
retrieving revision 1.3
diff -u -r1.3 netatalk.in
--- net/netatalk/files/netatalk.in	27 Mar 2010 00:13:49 -0000	1.3
+++ net/netatalk/files/netatalk.in	10 Oct 2011 06:42:51 -0000
@@ -1,9 +1,9 @@
 #!/bin/sh
 #
-# $FreeBSD: ports/net/netatalk/files/netatalk.in,v 1.3 2010/03/27 00:13:49 dougb Exp $
+# $FreeBSD: ports/net/netatalk/files/netatalk.in,v 1.4 2011/08/14 17:17:28 marcus Exp $
 #
 # PROVIDE: atalkd papd cnid_metad timelord afpd
-# REQUIRE: DAEMON %%SRVLOC%%
+# REQUIRE: DAEMON %%SRVLOC%% %%ZEROCONF%%
 # KEYWORD: shutdown
 #
 # AppleTalk daemons. Make sure not to start atalkd in the background:
Index: net/netatalk/files/patch-config_netatalk.conf
===================================================================
RCS file: /home/ncvs/ports/net/netatalk/files/patch-config_netatalk.conf,v
retrieving revision 1.3
diff -u -r1.3 patch-config_netatalk.conf
--- net/netatalk/files/patch-config_netatalk.conf	30 May 2010 17:03:17 -0000	1.3
+++ net/netatalk/files/patch-config_netatalk.conf	10 Oct 2011 06:42:51 -0000
@@ -1,43 +1,20 @@
---- config/netatalk.conf.orig	2010-04-19 11:35:26.000000000 +0000
-+++ config/netatalk.conf	2010-05-20 19:14:32.000000000 +0000
+--- config/netatalk.conf.orig	2011-07-22 00:30:42.000000000 -0400
++++ config/netatalk.conf	2011-07-30 18:01:04.000000000 -0400
 @@ -1,4 +1,5 @@
 -# Netatalk configuration
 +# netatalk configuration
 +# For details see man netatalk.conf
- # Change this to increase the maximum number of clients that can connect:
- AFPD_MAX_CLIENTS=20
  
-@@ -6,7 +7,7 @@
- # NOTE: if your zone has spaces in it, you're better off specifying
- #       it in afpd.conf
- #ATALK_ZONE=@zone
--ATALK_NAME=`echo ${HOSTNAME}|cut -d. -f1`
-+ATALK_NAME=`/bin/hostname -s`
+ #########################################################################
+ # Global configuration
+@@ -21,8 +22,8 @@ export ATALK_MAC_CHARSET
  
- # specify the Mac and unix charsets to be used
- ATALK_MAC_CHARSET='MAC_ROMAN'
-@@ -20,24 +21,5 @@
- # Change this to set the id of the guest user
- AFPD_GUEST=nobody
- 
--# Set which daemons to run.
--# If you need legacy AppleTalk, run atalkd.
--# papd, timelord and a2boot are dependent upon atalkd.
--# If you use "AFP over TCP" server only, run only cnid_metad and afpd.
--ATALKD_RUN=no
--PAPD_RUN=no
--TIMELORD_RUN=no
--A2BOOT_RUN=no
+ #### Set which daemons to run.
+ #### If you use AFP file server, run both cnid_metad and afpd.
 -CNID_METAD_RUN=yes
 -AFPD_RUN=yes
--
--# Control whether the daemons are started in the background.
--# If it is dissatisfied that atalkd starts slowly, set "yes".
--ATALK_BGROUND=no
--
--# export the charsets, read form ENV by apps
--export ATALK_MAC_CHARSET
--export ATALK_UNIX_CHARSET
--
- # config for cnid_metad. Default log config:
- # CNID_CONFIG="-l log_note"
++#CNID_METAD_RUN=yes
++#AFPD_RUN=yes
+ 
+ #### maximum number of clients that can connect:
+ #AFPD_MAX_CLIENTS=20
Index: net/netatalk/files/patch-configure
===================================================================
RCS file: /home/ncvs/ports/net/netatalk/files/patch-configure,v
retrieving revision 1.17
diff -u -r1.17 patch-configure
--- net/netatalk/files/patch-configure	25 Jul 2010 01:37:08 -0000	1.17
+++ net/netatalk/files/patch-configure	10 Oct 2011 06:42:51 -0000
@@ -1,11 +1,11 @@
---- configure.orig	2010-07-10 06:18:22.000000000 -0400
-+++ configure	2010-07-24 21:33:50.000000000 -0400
-@@ -30958,7 +30958,7 @@ if test "x$bdb_required" = "xyes"; then
+--- configure.orig	2011-07-27 07:59:28.000000000 -0400
++++ configure	2011-08-01 03:26:25.000000000 -0400
+@@ -31357,7 +31360,7 @@ if test "x$bdb_required" = "xyes"; then
      trybdbdir=""
      dobdbsearch=yes
      bdb_search_dirs="/usr/local /usr"
--    search_subdirs="/ /db5 /db5.0 /db50 /db4.8 /db48 /db4.7 /db47 /db4.6 /db46 /db4"
-+    search_subdirs="/%%DB_NAME%% / /db5 /db5.0 /db50 /db4.8 /db48 /db4.7 /db47 /db4.6 /db46 /db4"
+-    search_subdirs="/ /db5 /db5.1 /db51 /db5.0 /db50 /db4.8 /db48 /db4.7 /db47 /db4.6 /db46 /db4"
++    search_subdirs="/%%DB_NAME%% / /db5 /db5.1 /db51 /db5.0 /db50 /db4.8 /db48 /db4.7 /db47 /db4.6 /db46 /db4"
  
      bdbfound=no
      savedcflags="$CFLAGS"
Index: net/netatalk/files/patch-etc_apfd_Makefile.in
===================================================================
RCS file: /home/ncvs/ports/net/netatalk/files/patch-etc_apfd_Makefile.in,v
retrieving revision 1.1
diff -u -r1.1 patch-etc_apfd_Makefile.in
--- net/netatalk/files/patch-etc_apfd_Makefile.in	8 May 2011 23:08:21 -0000	1.1
+++ net/netatalk/files/patch-etc_apfd_Makefile.in	10 Oct 2011 06:42:51 -0000
@@ -1,11 +1,11 @@
---- etc/afpd/Makefile.in.orig	2011-05-08 19:04:22.000000000 -0400
-+++ etc/afpd/Makefile.in	2011-05-08 19:04:34.000000000 -0400
-@@ -295,7 +295,7 @@ afpd_SOURCES = unix.c ofork.c main.c swi
- 	afp_dsi.c messages.c afp_config.c nfsquota.c quota.c uam.c \
- 	afs.c uid.c afp_util.c catsearch.c afprun.c hash.c extattrs.c \
- 	$(am__append_1)
--afpd_LDADD = $(top_builddir)/libatalk/cnid/libcnid.la $(top_builddir)/libatalk/libatalk.la @QUOTA_LIBS@ @SLP_LIBS@ @WRAP_LIBS@ @LIBADD_DL@
-+afpd_LDADD = $(top_builddir)/libatalk/cnid/libcnid.la $(top_builddir)/libatalk/libatalk.la @QUOTA_LIBS@ @SLP_LIBS@ @WRAP_LIBS@ @LIBADD_DL@ @PAM_LIBS@
+--- etc/afpd/Makefile.in.orig	2011-07-27 07:59:24.000000000 -0400
++++ etc/afpd/Makefile.in	2011-07-30 18:03:30.000000000 -0400
+@@ -317,7 +317,7 @@ afpd_SOURCES = afp_asp.c afp_avahi.c afp
+ afpd_LDADD = \
+ 	$(top_builddir)/libatalk/cnid/libcnid.la \
+ 	$(top_builddir)/libatalk/libatalk.la \
+-	@LIBGCRYPT_LIBS@ @ZEROCONF_LIBS@ @QUOTA_LIBS@ @SLP_LIBS@ @WRAP_LIBS@ @LIBADD_DL@ @ACL_LIBS@ @PTHREAD_LIBS@
++	@LIBGCRYPT_LIBS@ @ZEROCONF_LIBS@ @QUOTA_LIBS@ @SLP_LIBS@ @WRAP_LIBS@ @LIBADD_DL@ @ACL_LIBS@ @PTHREAD_LIBS@ @PAM_LIBS@
+ 
  afpd_LDFLAGS = -export-dynamic 
- afpd_CFLAGS = -I$(top_srcdir)/include -I$(top_srcdir)/sys \
- 	 @SLP_CFLAGS@ \
+ afpd_CFLAGS = \
Index: net/netatalk/files/patch-etc_cnid_dbd_cnid_metad.c
===================================================================
RCS file: net/netatalk/files/patch-etc_cnid_dbd_cnid_metad.c
diff -N net/netatalk/files/patch-etc_cnid_dbd_cnid_metad.c
--- /dev/null	1 Jan 1970 00:00:00 -0000
+++ net/netatalk/files/patch-etc_cnid_dbd_cnid_metad.c	10 Oct 2011 06:42:51 -0000
@@ -0,0 +1,10 @@
+--- etc/cnid_dbd/cnid_metad.c.orig	2011-07-30 18:28:32.000000000 -0400
++++ etc/cnid_dbd/cnid_metad.c	2011-07-30 18:29:01.000000000 -0400
+@@ -39,6 +39,7 @@
+ #include <string.h>
+ #include <signal.h>
+ #include <sys/types.h>
++#include <sys/resource.h>
+ #include <sys/time.h>
+ #include <sys/wait.h>
+ #include <sys/uio.h>
Index: net/netatalk/files/patch-include_atalk_dsi.h
===================================================================
RCS file: net/netatalk/files/patch-include_atalk_dsi.h
diff -N net/netatalk/files/patch-include_atalk_dsi.h
--- net/netatalk/files/patch-include_atalk_dsi.h	9 Jun 2010 22:49:30 -0000	1.1
+++ /dev/null	1 Jan 1970 00:00:00 -0000
@@ -1,10 +0,0 @@
---- include/atalk/dsi.h.orig	2010-06-09 18:46:15.000000000 -0400
-+++ include/atalk/dsi.h	2010-06-09 18:44:22.000000000 -0400
-@@ -11,6 +11,7 @@
- #include <sys/time.h>
- #include <signal.h>
- 
-+#include <sys/socket.h>
- #include <netinet/in.h>
- #include <atalk/afp.h>
- #include <atalk/server_child.h>
