diff --git sysutils/megacli/Makefile sysutils/megacli/Makefile
index 2a039b4..78a81fa 100644
--- sysutils/megacli/Makefile
+++ sysutils/megacli/Makefile
@@ -2,11 +2,11 @@
 # Date created:		12 April 2009
 # Whom:			Vladimir Laskov <samflanker@gmail.com>
 #
-# $FreeBSD: ports/sysutils/megacli/Makefile,v 1.12 2011/11/30 15:30:11 pawel Exp $
+# $FreeBSD: head/sysutils/megacli/Makefile 308592 2012-12-10 08:48:38Z delphij $
 #
 
 PORTNAME=	megacli
-PORTVERSION=	8.02.16
+PORTVERSION=	8.05.06
 CATEGORIES=	sysutils
 MASTER_SITES=	http://www.lsi.com/downloads/Public/MegaRAID%20Common%20Files/ \
 		CRITICAL
@@ -28,13 +28,20 @@ NO_WRKSUBDIR=	yes
 PLIST_FILES=	sbin/MegaCli
 
 pre-fetch:
-	@${FETCH_BINARY} -o /dev/null http://www.lsi.com/magic.axd?x=e&file=http%3A//www.lsi.com/downloads/Public/MegaRAID%2520Common%2520Files/8.02.16_MegaCLI.zip
+	@${FETCH_BINARY} -o /dev/null http://www.lsi.com/magic.axd?x=e&file=http%3A//www.lsi.com/downloads/Public/MegaRAID%2520Common%2520Files/${PORT_VERSION}_MegaCLI.zip
 
 .include <bsd.port.pre.mk>
+post-extract:
+.if ${ARCH} == "i386"
+	@cd ${WRKSRC}/MegaCli_FreeBSD && ${UNZIP_CMD} ${EXTRACT_BEFORE_ARGS} MegaCli.zip
+.elif ${ARCH} == "amd64"
+	@cd ${WRKSRC}/MegaCli_FreeBSD && ${UNZIP_CMD} ${EXTRACT_BEFORE_ARGS} MegaCli64.zip
+.endif
+
 do-install:
 .if ${ARCH} == "i386"
-		${INSTALL_PROGRAM} ${WRKSRC}/FREEBSD/MegaCli ${PREFIX}/sbin/MegaCli
+		${INSTALL_PROGRAM} ${WRKSRC}/MegaCli_FreeBSD/MegaCli ${PREFIX}/sbin/MegaCli
 .elif ${ARCH} == "amd64"
-		${INSTALL_PROGRAM} ${WRKSRC}/FREEBSD/MegaCli64 ${PREFIX}/sbin/MegaCli
+		${INSTALL_PROGRAM} ${WRKSRC}/MegaCli_FreeBSD/MegaCli64 ${PREFIX}/sbin/MegaCli
 .endif
 .include <bsd.port.post.mk>
diff --git sysutils/megacli/distinfo sysutils/megacli/distinfo
index 51aa286..2172bb2 100644
--- sysutils/megacli/distinfo
+++ sysutils/megacli/distinfo
@@ -1,2 +1,2 @@
-SHA256 (8.02.16_MegaCLI.zip) = e22b615f6a10e4c305e65434153f8e15e779a2e245c08116ce5eca840c183d66
-SIZE (8.02.16_MegaCLI.zip) = 12244704
+SHA256 (8.05.06_MegaCLI.zip) = 7b310559b5e3a07cb02da6280e98ad4a95af258153ad0487db8c6187713e0f45
+SIZE (8.05.06_MegaCLI.zip) = 18874438
