--- net/istgt/Makefile	2011-08-19 14:30:30.000000000 -0700
+++ net/istgt/Makefile	2011-08-19 14:32:10.000000000 -0700
@@ -2,11 +2,11 @@
 # Date created:		2009-02-23
 # Whom:			Daisuke Aoyama <aoyama@peach.ne.jp>
 #
-# $FreeBSD: ports/net/istgt/Makefile,v 1.23 2011/05/30 20:10:18 wxs Exp $
+# $FreeBSD: ports/net/istgt/Makefile,v 1.24 2011/08/08 07:11:07 culot Exp $
 #
 
 PORTNAME=	istgt
-PORTVERSION=	20110529
+PORTVERSION=	20110717
 CATEGORIES=	net
 MASTER_SITES=	http://www.peach.ne.jp/archives/istgt/
 
@@ -17,6 +17,7 @@
 PKGMESSAGE=	${WRKDIR}/pkg-message
 
 GNU_CONFIGURE=	yes
+CONFIGURE_ARGS+=	--with-aio
 
 MAN1=		istgt.1
 MANCOMPRESSED=	yes
--- net/istgt/distinfo	2011-08-19 14:30:30.000000000 -0700
+++ net/istgt/distinfo	2011-08-19 14:31:07.000000000 -0700
@@ -1,2 +1,2 @@
-SHA256 (istgt-20110529.tar.gz) = 51a3bde2177780f1a9cd0b32d3949f5db5e614f8d2c520e408e329778f7c49e0
-SIZE (istgt-20110529.tar.gz) = 224527
+SHA256 (istgt-20110717.tar.gz) = b9605c6ca4286e25a581791871ca4c90d253b4aad6d25bbde6eed0713d50bdb3
+SIZE (istgt-20110717.tar.gz) = 229746
