--- net/istgt/Makefile	2011-05-30 13:10:18.000000000 -0700
+++ net/istgt/Makefile	2011-09-18 09:46:41.000000000 -0700
@@ -2,11 +2,11 @@
 # Date created:		2009-02-23
 # Whom:			Daisuke Aoyama <aoyama@peach.ne.jp>
 #
-# $FreeBSD: ports/net/istgt/Makefile,v 1.23 2011/05/30 20:10:18 wxs Exp $
+# $FreeBSD: ports/net/istgt/Makefile,v 1.25 2011/09/07 17:25:49 sunpoet Exp $
 #
 
 PORTNAME=	istgt
-PORTVERSION=	20110529
+PORTVERSION=	20110918
 CATEGORIES=	net
 MASTER_SITES=	http://www.peach.ne.jp/archives/istgt/
 
@@ -17,6 +17,7 @@
 PKGMESSAGE=	${WRKDIR}/pkg-message
 
 GNU_CONFIGURE=	yes
+CONFIGURE_ARGS+=	--with-aio
 
 MAN1=		istgt.1
 MANCOMPRESSED=	yes
--- net/istgt/distinfo	2011-09-18 09:47:05.000000000 -0700
+++ net/istgt/distinfo	2011-09-18 09:46:45.000000000 -0700
@@ -1,2 +1,2 @@
-SHA256 (istgt-20110529.tar.gz) = 51a3bde2177780f1a9cd0b32d3949f5db5e614f8d2c520e408e329778f7c49e0
-SIZE (istgt-20110529.tar.gz) = 224527
+SHA256 (istgt-20110918.tar.gz) = 00f2d9958437212d5a0131e06d1ea9d9f27daa673a56cc94abc134e2a555fed4
+SIZE (istgt-20110918.tar.gz) = 231629
--- net/istgt/files/patch-etc__istgt.sh.in	1969-12-31 16:00:00.000000000 -0800
+++ net/istgt/files/patch-etc__istgt.sh.in	2011-09-18 09:45:10.000000000 -0700
@@ -0,0 +1,10 @@
+--- ./etc/istgt.sh.in.orig	2011-05-29 03:53:29.000000000 -0700
++++ ./etc/istgt.sh.in	2011-08-23 11:14:24.910003136 -0700
+@@ -2,7 +2,6 @@
+ 
+ # PROVIDE: istgt
+ # REQUIRE: NETWORKING
+-# BEFORE: mountcritremote
+ # KEYWORD: shutdown
+ 
+ . /etc/rc.subr
