--- devel/py-polib/Makefile	17 Aug 2009 08:10:35 -0000	1.6
+++ devel/py-polib/Makefile	14 Mar 2012 09:45:59 -0000
@@ -6,9 +6,9 @@
 #
 
 PORTNAME=	polib
-PORTVERSION=	0.4.1
+PORTVERSION=	0.7.0
 CATEGORIES=	devel python
-MASTER_SITES=	${MASTER_SITE_GOOGLE_CODE}
+MASTER_SITES=	http://cdn.bitbucket.org/izi/polib/downloads/
 PKGNAMEPREFIX=	${PYTHON_PKGNAMEPREFIX}
 
 MAINTAINER=	yzlin@FreeBSD.org
--- devel/py-polib/distinfo	3 Jul 2011 14:58:45 -0000	1.4
+++ devel/py-polib/distinfo	14 Mar 2012 09:46:04 -0000
@@ -1,2 +1,2 @@
-SHA256 (polib-0.4.1.tar.gz) = 37513d2edbb1f5da85df9daa27ffc16c0e15a7c49283990fbf15cb378e8e8d78
-SIZE (polib-0.4.1.tar.gz) = 211914
+SHA256 (polib-0.7.0.tar.gz) = d9184d7454c650f5a26db54e70da8a50c51dcc6873af65277708ab9c6cf5c449
+SIZE (polib-0.7.0.tar.gz) = 149614
