Index: net-p2p/transmission-cli/Makefile
===================================================================
RCS file: /home/ncvs/ports/net-p2p/transmission-cli/Makefile,v
retrieving revision 1.86
diff -u -r1.86 Makefile
--- net-p2p/transmission-cli/Makefile	1 Jun 2012 05:23:22 -0000	1.86
+++ net-p2p/transmission-cli/Makefile	10 Sep 2012 13:05:03 -0000
@@ -2,14 +2,14 @@
 # Date created:				08 October 2005
 # Whom:					Mezz <mezz@FreeBSD.org>
 #
-# $FreeBSD: ports/net-p2p/transmission-cli/Makefile,v 1.86 2012/06/01 05:23:22 dinoex Exp $
+# $FreeBSD: ports/net-p2p/transmission-cli/Makefile,v 1.88 2012/08/10 19:59:17 crees Exp $
 #
 
 PORTNAME=	transmission
-PORTVERSION=	2.51
-PORTREVISION?=	1
+PORTVERSION=	2.61
+PORTREVISION?=	0
 CATEGORIES?=	net-p2p
-#MASTER_SITES=	http://download.m0k.org/${PORTNAME}/files/ \
+#MASTER_SITES=	http://download.m0k.org/${PORTNAME}/files/
 MASTER_SITES=	http://transmission.cachefly.net/
 PKGNAMESUFFIX?=	-cli
 
@@ -17,6 +17,9 @@
 COMMENT?=	A lightweight CLI (command line) BitTorrent client and utilities
 
 SLAVEPORT?=	cli
+
+CONFLICTS_INSTALL=	${PORTNAME}${PKGNAMESUFFIX}-2.5*
+
 .if ${SLAVEPORT}!="web"
 LIB_DEPENDS?=	${GEN_LIB_DEPENDS}
 
Index: net-p2p/transmission-cli/distinfo
===================================================================
RCS file: /home/ncvs/ports/net-p2p/transmission-cli/distinfo,v
retrieving revision 1.60
diff -u -r1.60 distinfo
--- net-p2p/transmission-cli/distinfo	17 Apr 2012 20:13:41 -0000	1.60
+++ net-p2p/transmission-cli/distinfo	10 Sep 2012 13:05:03 -0000
@@ -1,2 +1,2 @@
-SHA256 (transmission-2.51.tar.xz) = 1c9a54864acf8fd0c7aa4132e9e69b454a7c72c305c96016a38551157af8858b
-SIZE (transmission-2.51.tar.xz) = 2784036
+SHA256 (transmission-2.61.tar.xz) = 1c4ccbf5d7dc7bb9615005f9db7bb0612c95ac908c6cf5d519e6f0126588765c
+SIZE (transmission-2.61.tar.xz) = 2876648
--- net-p2p/transmission-cli/files/patch-third-party_libutp_utp.cpp	2012-09-10 10:02:08.672762175 -0300
+++ /dev/null	2012-09-10 10:00:59.000000000 -0300
@@ -1,12 +0,0 @@
---- third-party/libutp/utp.cpp.orig	2011-04-29 21:45:38.000000000 -0500
-+++ third-party/libutp/utp.cpp	2011-04-29 21:47:05.000000000 -0500
-@@ -10,6 +10,9 @@
- #include <stdlib.h>
- #include <errno.h>
- #include <limits.h> // for UINT_MAX
-+#include <sys/types.h>
-+#include <sys/socket.h>
-+#include <netinet/in.h>
- 
- #ifdef WIN32
- #include "win32_inet_ntop.h"
