--- devel/gobject-introspection/Makefile~	2011-11-20 13:42:24.555693523 -0800
+++ devel/gobject-introspection/Makefile	2011-11-20 13:42:40.575690618 -0800
@@ -8,6 +8,7 @@
 
 PORTNAME=	gobject-introspection
 PORTVERSION=	0.10.8
+PORTREVISION=	1
 CATEGORIES=	devel
 MASTER_SITES=	GNOME
 DIST_SUBDIR=	gnome3
@@ -15,8 +16,7 @@
 MAINTAINER=	gnome@FreeBSD.org
 COMMENT=	Generate interface introspection data for GObject libraries
 
-LIB_DEPENDS=	ffi.5:${PORTSDIR}/devel/libffi \
-		cairo.2:${PORTSDIR}/graphics/cairo
+LIB_DEPENDS=	ffi.5:${PORTSDIR}/devel/libffi
 
 USE_BZIP2=	yes
 GNU_CONFIGURE=	yes
