diff --exclude '*.orig' -Naur dns/inadyn/Makefile dns/inadyn/Makefile
--- dns/inadyn/Makefile	29 Jan 2012 10:56:06 -0000	1.10
+++ dns/inadyn/Makefile	13 Jul 2012 11:35:37 -0000
@@ -21,35 +21,27 @@
 USE_RC_SUBR=	inadyn
 USE_ZIP=	yes
 USE_DOS2UNIX=	yes
-USE_GMAKE=	yes
-MAKEFILE=	makefile
-
-.if !defined(NO_INSTALL_MANPAGES)
-MAN8=		inadyn.8
-MAN5=		inadyn.conf.5
-.endif
 
 PORTDOCS=	readme.html
 PLIST_FILES=	bin/inadyn
 
+MAKE_JOBS_SAFE=        yes
+MAKEFILE=      ${FILESDIR}/BSDmakefile
+
+MAKE_ENV+=     MANDIR=${MANPREFIX}/man
+
 post-extract:
 	@${MV} ${WRKDIR}/${PORTNAME} ${WRKDIR}/${PORTNAME}-${PORTVERSION}
 
 post-patch:
-	@${REINPLACE_CMD} -e 's|gcc|$$\(CC\)|g' ${WRKSRC}/makefile
 	@${REINPLACE_CMD} -e "s,%%PREFIX%%,${PREFIX},g" \
 		${WRKSRC}/man/inadyn.8 ${WRKSRC}/readme.html \
 		${WRKSRC}/src/dyndns.h
 
-do-install:
+pre-install:
 .if !defined(NOPORTDOCS)
 	${MKDIR} ${DOCSDIR}
 	${INSTALL_DATA} -p ${WRKSRC}/${PORTDOCS} ${DOCSDIR}
 .endif
-	${INSTALL_PROGRAM} ${WRKSRC}/bin/linux/inadyn ${PREFIX}/bin
-.if !defined(NO_INSTALL_MANPAGES)
-	${INSTALL_MAN} ${WRKSRC}/man/inadyn.8 ${MANPREFIX}/man/man8
-	${INSTALL_MAN} ${WRKSRC}/man/inadyn.conf.5 ${MANPREFIX}/man/man5
-.endif
 
 .include <bsd.port.mk>
diff --exclude '*.orig' -Naur dns/inadyn/files/BSDmakefile dns/inadyn/files/BSDmakefile
--- dns/inadyn/files/BSDmakefile	1969-12-31 16:00:00.000000000 -0800
+++ dns/inadyn/files/BSDmakefile	2012-01-07 20:02:28.322979665 -0800
@@ -0,0 +1,16 @@
+PROG=	inadyn
+
+MAN5=	inadyn.conf.5
+
+MAN8=	inadyn.8
+
+SRCS=	base64utils.c dyndns.c errorcode.c get_cmd.c http_client.c ip.c main.c os.c \
+	os_psos.c os_unix.c os_windows.c tcp.c inadyn_cmd.c
+
+.PATH:	man src
+
+WARNS=	2
+
+BINDIR=	${PREFIX}/bin
+
+.include <bsd.prog.mk>
