Index: sysutils/tw_cli/Makefile
===================================================================
--- sysutils/tw_cli/Makefile	(revision 300790)
+++ sysutils/tw_cli/Makefile	(working copy)
@@ -8,7 +8,7 @@
 PORTNAME=	tw_cli
 PORTVERSION=	9.5.4
 CATEGORIES=	sysutils
-MASTER_SITES=	ftp://hosting.lissyara.su/
+MASTER_SITES=	http://www.lsi.com/downloads/Public/SATA/SATA%20Common%20Files/
 DISTNAME=	cli_freebsd_10.2.1_9.5.4
 EXTRACT_SUFX=	.zip
 
