Index: ftp/proftpd/Makefile
===================================================================
RCS file: /home/ncvs/ports/ftp/proftpd/Makefile,v
retrieving revision 1.195
diff -u -p -r1.195 ftp/proftpd/Makefile
--- ftp/proftpd/Makefile	10 Nov 2011 23:28:21 -0000	1.195
+++ ftp/proftpd/Makefile	16 Dec 2011 23:17:52 -0000
@@ -7,6 +7,7 @@
 
 PORTNAME=	proftpd
 PORTVERSION=	1.3.3e
+PORTREVISION=	1
 CATEGORIES?=	ftp
 MASTER_SITES=	ftp://ftp.proftpd.org/distrib/source/ \
 		ftp://ftp.fastorama.com/mirrors/ftp.proftpd.org/distrib/source/ \
Index: ftp/proftpd/files/patch-src-fsio.c
===================================================================
RCS file: files/patch-src-fsio.c
diff -N ftp/proftpd/files/patch-src-fsio.c
--- /dev/null	1 Jan 1970 00:00:00 -0000
+++ ftp/proftpd/files/patch-src-fsio.c	16 Dec 2011 23:12:28 -0000
@@ -0,0 +1,19 @@
+--- src/fsio.c.orig	2010-04-12 12:00:00.000000000 -0700
++++ src/fsio.c	2011-12-16 15:12:07.799166185 -0800
+@@ -50,6 +50,8 @@
+ # include <acl/libacl.h>
+ #endif
+ 
++#include <unistd.h>
++
+ typedef struct fsopendir fsopendir_t;
+ 
+ struct fsopendir {
+@@ -287,6 +289,7 @@
+ static int sys_chroot(pr_fs_t *fs, const char *path) {
+   if (chroot(path) < 0)
+     return -1;
++  __FreeBSD_libc_enter_restricted_mode();
+ 
+   session.chroot_path = (char *) path;
+   return 0;
