Index: net/istgt/Makefile
===================================================================
RCS file: /home/ncvs/ports/net/istgt/Makefile,v
retrieving revision 1.21
diff -u -p -r1.21 Makefile
--- net/istgt/Makefile	23 Feb 2011 18:09:53 -0000	1.21
+++ net/istgt/Makefile	1 Apr 2011 17:54:56 -0000
@@ -7,6 +7,7 @@
 
 PORTNAME=	istgt
 PORTVERSION=	20110223
+PORTREVISION=	1
 CATEGORIES=	net
 MASTER_SITES=	http://www.peach.ne.jp/archives/istgt/
 
Index: net/istgt/files/patch-etc__istgt.sh.in
===================================================================
RCS file: files/patch-etc__istgt.sh.in
diff -N files/patch-etc__istgt.sh.in
--- /dev/null	1 Jan 1970 00:00:00 -0000
+++ net/istgt/files/patch-etc__istgt.sh.in	1 Apr 2011 17:54:29 -0000
@@ -0,0 +1,10 @@
+--- ./etc/istgt.sh.in.orig	2009-04-20 02:46:41.000000000 -0700
++++ ./etc/istgt.sh.in	2011-04-01 10:54:21.980657627 -0700
+@@ -9,6 +9,7 @@
+ 
+ name="istgt"
+ rcvar=`set_rcvar`
++extra_commands="reload"
+ 
+ load_rc_config $name
+ 
