Index: sysutils/testdisk/Makefile
===================================================================
RCS file: /home/ncvs/ports/sysutils/testdisk/Makefile,v
retrieving revision 1.29
diff -u -r1.29 Makefile
--- sysutils/testdisk/Makefile	16 Feb 2012 04:44:32 -0000	1.29
+++ sysutils/testdisk/Makefile	1 Apr 2013 11:37:55 -0000
@@ -6,10 +6,11 @@
 #
 
 PORTNAME=	testdisk
-PORTVERSION=	6.13
+PORTVERSION=	6.14
 CATEGORIES=	sysutils
 MASTER_SITES=	http://www.cgsecurity.org/
-DISTFILES=	${PORTNAME}-${PORTVERSION}${EXTRACT_SUFX}
+DISTVERSIONSUFFIX=	-WIP
+DISTFILES=	${PORTNAME}-${PORTVERSION}-WIP${EXTRACT_SUFX}
 
 MAINTAINER=	flz@FreeBSD.org
 COMMENT=	Tool to check and undelete partitions
@@ -53,7 +54,7 @@
 #.endif
 
 .if defined(WITH_NTFS3G)
-LIB_DEPENDS+=	ntfs-3g.81:${PORTSDIR}/sysutils/fusefs-ntfs
+LIB_DEPENDS+=	ntfs-3g.83:${PORTSDIR}/sysutils/fusefs-ntfs
 .else
 CONFIGURE_ARGS+=--without-ntfs3g
 .endif
Index: sysutils/testdisk/distinfo
===================================================================
RCS file: /home/ncvs/ports/sysutils/testdisk/distinfo,v
retrieving revision 1.18
diff -u -r1.18 distinfo
--- sysutils/testdisk/distinfo	16 Feb 2012 04:44:32 -0000	1.18
+++ sysutils/testdisk/distinfo	1 Apr 2013 11:37:55 -0000
@@ -1,2 +1,2 @@
-SHA256 (testdisk-6.13.tar.bz2) = 48275ccddfc9445b6c3e6e181728dae9e49bd25b824a6b5670d58f1f88cdf220
-SIZE (testdisk-6.13.tar.bz2) = 560696
+SHA256 (testdisk-6.14-WIP.tar.bz2) = f3362dff045c6d5117ea65471ac2ce359ef9210d2c6251764a7e1daa6c0a5e3d
+SIZE (testdisk-6.14-WIP.tar.bz2) = 580464
