diff --git net/samba36/Makefile net/samba36/Makefile
index 2f85bc5..e5829bc 100644
--- net/samba36/Makefile
+++ net/samba36/Makefile
@@ -2,11 +2,11 @@
 # Date created:				1 May 2011
 # Whom:					timur@FreeBSD.org
 #
-# $FreeBSD: head/net/samba36/Makefile 306898 2012-11-03 03:08:45Z timur $
+# $FreeBSD: head/net/samba36/Makefile 311317 2013-01-31 15:59:44Z timur $
 #
 
 PORTNAME=		${SAMBA_BASENAME}36
-PORTVERSION=		3.6.9
+PORTVERSION=		3.6.12
 PORTREVISION?=		0
 CATEGORIES?=		net
 MASTER_SITES=		${MASTER_SITE_SAMBA}
@@ -65,8 +65,8 @@ CONFIGURE_ARGS+=	--exec-prefix="${PREFIX}" \
 			--datadir="${DATADIR}" \
 			--with-swatdir="${SAMBA_SWATDIR}" \
 			--libdir="${SAMBA_LIBDIR}" \
-			--with-modulesdir="${SAMBA_MODULEDIR}" \
 			--with-pammodulesdir="${SAMBA_LIBDIR}" \
+			--with-modulesdir="${SAMBA_MODULEDIR}" \
 			--localstatedir="${VARDIR}" \
 			--with-piddir="${SAMBA_RUNDIR}" \
 			--with-ncalrpcdir="${SAMBA_RUNDIR}/ncalrpc" \
@@ -116,7 +116,7 @@ OPTIONS=	LDAP		"With LDAP support" on \
 		PAM_SMBPASS	"With PAM authentication vs passdb backends" off \
 		DNSUPDATE	"With dynamic DNS update(require ADS)" off \
 		AVAHI		"With Bonjour service discovery support" off \
-		PTHREADPOOL	"With pthread pool" off \
+		PTHREADPOOL	"With pthread pool" on \
 		EXP_MODULES	"With experimental modules" off \
 		POPT		"With system-wide POPT library" on \
 		IPV6		"With IPv6 support" on \
@@ -184,10 +184,10 @@ SUB_LIST+=		CUPS="@comment "
 CPPFLAGS+=		-g
 LDFLAGS+=		-g
 LIB_DEPENDS+=		dmalloc.1:${PORTSDIR}/devel/dmalloc
-CONFIGURE_ARGS+=	--enable-debug \
-			--enable-socket-wrapper --enable-nss-wrapper \
-			--enable-developer --enable-krb5developer \
-			--enable-dmalloc --with-profiling-data
+CONFIGURE_ARGS+=	--enable-debug
+#			--enable-socket-wrapper --enable-nss-wrapper \
+#			--enable-developer --enable-krb5developer \
+#			--enable-dmalloc --with-profiling-data
 
 CONFIGURE_ARGS+=	--with-smbtorture4-path=${WRKDIR}/${DISTNAME}/source4/torture
 .else
diff --git net/samba36/distinfo net/samba36/distinfo
index d438c20..79cde8b 100644
--- net/samba36/distinfo
+++ net/samba36/distinfo
@@ -1,2 +1,2 @@
-SHA256 (samba-3.6.9.tar.gz) = 9714b50c1bbdb0d3521bdbf403bb112b5405040685cb0b4ff3f9a370d1e84020
-SIZE (samba-3.6.9.tar.gz) = 34077002
+SHA256 (samba-3.6.12.tar.gz) = 6ff797c1772613bd5172d2c8b900fd56bf096d7252faee8b47c4925a4fdc9f8b
+SIZE (samba-3.6.12.tar.gz) = 34073788
diff --git net/samba36/files/patch-lib__tevent__tevent_signal.c net/samba36/files/patch-lib__tevent__tevent_signal.c
index 8d59153..d62b2e6 100644
--- net/samba36/files/patch-lib__tevent__tevent_signal.c
+++ net/samba36/files/patch-lib__tevent__tevent_signal.c
@@ -5,7 +5,7 @@
  #include "tevent_util.h"
  
 -#define TEVENT_NUM_SIGNALS 64
-+#define TEVENT_NUM_SIGNALS SIGRTMAX
++#define TEVENT_NUM_SIGNALS (SIGRTMAX+1)
  
  /* maximum number of SA_SIGINFO signals to hold in the queue.
    NB. This *MUST* be a power of 2, in order for the ring buffer
diff --git net/samba36/files/patch-source3__auth__user_util.c net/samba36/files/patch-source3__auth__user_util.c
new file mode 100644
index 0000000..430fea3
--- /dev/null
+++ net/samba36/files/patch-source3__auth__user_util.c
@@ -0,0 +1,22 @@
+--- ./source3/auth/user_util.c.orig	2012-09-14 08:12:09.000000000 +0000
++++ ./source3/auth/user_util.c	2012-10-18 20:58:57.065810414 +0000
+@@ -425,12 +425,14 @@
+ 	x_fclose(f);
+ 
+ 	/*
+-	 * Setup the last_from and last_to as an optimization so
+-	 * that we don't scan the file again for the same user.
++	 * If mapping failed setup the last_from and last_to as
++	 * an optimization so that we don't scan the file again
++	 * for the same user.
+ 	 */
+-
+-	set_last_from_to(user_in, user_in);
+-	store_map_in_gencache(ctx, user_in, user_in);
++	if(!mapped_user) {
++		set_last_from_to(user_in, user_in);
++		store_map_in_gencache(ctx, user_in, user_in);
++	}
+ 
+ 	return mapped_user;
+ }
diff --git net/samba36/files/patch-source3__configure.in net/samba36/files/patch-source3__configure.in
index d2433da..30a50f9 100644
--- net/samba36/files/patch-source3__configure.in
+++ net/samba36/files/patch-source3__configure.in
@@ -1,26 +1,6 @@
---- ./source3/configure.in.orig	2012-10-29 09:05:46.000000000 +0000
-+++ ./source3/configure.in	2012-11-01 13:16:20.589876378 +0000
-@@ -759,7 +759,6 @@
- if test x"$ac_cv_header_md5_h" = x"yes"; then
- 	AC_DEFINE(HAVE_MD5_H, 1,
- 		[Whether md5.h is available.])
--	AC_CHECK_LIB(md5, MD5Update, [samba_cv_md5lib=md5])
- fi
- 
- if test x"$ac_cv_header_md5_h" = x"yes" -a \
-@@ -769,6 +768,11 @@
- 
- if test x"$ac_cv_header_md5_h" = x"yes" -a \
-         x"$samba_cv_md5lib" = x"none" ; then
-+	AC_CHECK_LIB(md5, MD5Update, [samba_cv_md5lib=md5])
-+fi
-+
-+if test x"$ac_cv_header_md5_h" = x"yes" -a \
-+        x"$samba_cv_md5lib" = x"none" ; then
- 	AC_CHECK_LIB(c, MD5Update, [samba_cv_md5lib=""])
- fi
- 
-@@ -868,9 +872,13 @@
+--- ./source3/configure.in.orig	2013-01-18 10:41:08.000000000 +0000
++++ ./source3/configure.in	2013-01-30 01:24:19.349878982 +0000
+@@ -864,9 +864,13 @@
  	AC_PATH_PROG(CUPS_CONFIG, cups-config)
  
          if test "x$CUPS_CONFIG" != x; then
@@ -35,7 +15,7 @@
  		AC_CHECK_HEADERS(cups/cups.h cups/language.h)
  
  		if test x"$ac_cv_header_cups_cups_h" = xyes -a \
-@@ -879,7 +887,7 @@
+@@ -875,7 +879,7 @@
  			# underlinked. With cups-config --libs we pull in unwanted and unneeded
  			# dendencies including thread libraries - use cups-config only if really
  			# required. 
@@ -44,7 +24,7 @@
  				[PRINT_LIBS="$ac_save_PRINT_LIBS -lcups"],
  				[AC_MSG_WARN([your cups library doesn't link with -lcups alone, it might be underlinked.]) ;
  				 PRINT_LIBS="$ac_save_PRINT_LIBS `$CUPS_CONFIG --libs`"])
-@@ -889,10 +897,10 @@
+@@ -885,10 +889,10 @@
  			AC_CHECK_LIB_EXT(cups, PRINT_LIBS, httpConnectEncrypt)
  		else
  			AC_MSG_WARN([cups-config around but cups-devel not installed])
@@ -57,7 +37,7 @@
  
  	elif test x"$enable_cups" = x"yes"; then
  		AC_MSG_ERROR(Cups support required but cups-config not located.  Make sure cups-devel related files are installed.)
-@@ -1849,7 +1857,7 @@
+@@ -1845,7 +1849,7 @@
  fi
  
  if test x"$BLDSHARED" = x"true" ; then
@@ -66,7 +46,7 @@
  fi
  
  AC_MSG_RESULT($BLDSHARED)
-@@ -1980,9 +1988,6 @@
+@@ -1976,9 +1980,6 @@
  AC_MSG_CHECKING([PICFLAG])
  AC_MSG_RESULT([$PICFLAG])
  
@@ -76,7 +56,7 @@
  AC_CACHE_CHECK([whether building shared libraries actually works],
                 [ac_cv_shlib_works],[
     # try building a trivial shared library
-@@ -4639,15 +4644,51 @@
+@@ -4656,15 +4657,51 @@
  		x"$ac_cv_header_pam_pam_modules_h" = x"no" ; then
  		if test x"${try_pam}" = x"yes";then
  			AC_MSG_ERROR([--with-pam=yes but pam_modules.h not found])
@@ -132,7 +112,7 @@
  
  		if test x"$create_pam_modules" = x"yes"; then
  			AC_DEFINE(WITH_PAM_MODULES,1,[Whether to include PAM MODULES support])
-@@ -4657,7 +4698,7 @@
+@@ -4674,7 +4711,7 @@
  			AC_CHECK_HEADERS(pam/pam_ext.h pam/_pam_macros.h)
  			AC_CHECK_FUNC_EXT(pam_vsyslog,$PAM_LIBS)
  		else
@@ -141,28 +121,16 @@
  		fi
  	fi
  	AC_MSG_CHECKING(whether to use PAM support)
-@@ -5859,9 +5900,17 @@
- [AC_MSG_RESULT(no)])
+@@ -5877,7 +5914,7 @@
  
  			AC_MSG_CHECKING(for aio_suspend)
--			AC_LINK_IFELSE([#include <aio.h>
--int main() { struct aiocb a; return aio_suspend(&a, 1, NULL); }],
--[AC_DEFINE(HAVE_AIO_SUSPEND, 1, [Have aio_suspend]) AC_MSG_RESULT(yes)],
-+			AC_LINK_IFELSE([
-+				#include <aio.h>
-+				#if STDC_HEADERS
-+				#include <stdlib.h>
-+				#include <stddef.h>
-+				#endif
-+				int main() {
-+					struct aiocb *a[1];
-+					return aio_suspend((const struct aiocb**)&a, 1, NULL);
-+				}
-+			], [AC_DEFINE(HAVE_AIO_SUSPEND, 1, [Have aio_suspend]) AC_MSG_RESULT(yes)],
+ 			AC_LINK_IFELSE([#include <aio.h>
+-int main() { struct aiocb a; struct timespec t; return aio_suspend(&a, 1, &t); }],
++int main() { struct aiocb *a[[1]]; struct timespec t; return aio_suspend((const struct aiocb**)&a, 1, &t); }],
+ [AC_DEFINE(HAVE_AIO_SUSPEND, 1, [Have aio_suspend]) AC_MSG_RESULT(yes)],
  [AC_MSG_RESULT(no)])
  		fi
- 
-@@ -6319,12 +6368,13 @@
+@@ -6336,12 +6373,13 @@
  		NSSSONAMEVERSIONSUFFIX=".2"
  		WINBIND_NSS_EXTRA_OBJS="../nsswitch/winbind_nss_linux.o"
  		;;
@@ -177,7 +145,7 @@
  		WINBIND_NSS="../nsswitch/nss_winbind.$SHLIBEXT"
  		WINBIND_WINS_NSS="../nsswitch/nss_wins.$SHLIBEXT"
  		;;
-@@ -6471,10 +6521,14 @@
+@@ -6488,10 +6526,14 @@
          AC_MSG_RESULT(no$winbind_no_reason)
  fi
  
@@ -192,7 +160,7 @@
  AC_SUBST(NSSSONAMEVERSIONSUFFIX)
  
  AC_SUBST(WINBIND_KRB5_LOCATOR)
-@@ -6793,13 +6847,21 @@
+@@ -6810,13 +6852,21 @@
  # Start
  AC_CHECK_FUNC(getmntent)
  
@@ -215,7 +183,7 @@
  		int main(void)
  		{
  			struct statfs fsd;
-@@ -7099,6 +7161,16 @@
+@@ -7116,6 +7166,16 @@
  	LIBS="$LIBS -ldmalloc"
  fi
  
@@ -232,7 +200,7 @@
  dnl Remove -L/usr/lib/? from LDFLAGS and LIBS
  LIB_REMOVE_USR_LIB(LDFLAGS)
  LIB_REMOVE_USR_LIB(LIBS)
-@@ -7147,6 +7219,7 @@
+@@ -7164,6 +7224,7 @@
  AC_SUBST(SMBD_LIBS)
  
  CFLAGS="${CFLAGS} \$(FLAGS)"
diff --git net/samba36/files/patch-source3__utils__net.c net/samba36/files/patch-source3__utils__net.c
new file mode 100644
index 0000000..aa4853a
--- /dev/null
+++ net/samba36/files/patch-source3__utils__net.c
@@ -0,0 +1,18 @@
+--- ./source3/utils/net.c.orig	2013-01-29 08:49:31.000000000 +0000
++++ ./source3/utils/net.c	2013-01-31 15:04:24.069877314 +0000
+@@ -898,8 +898,13 @@
+ 		set_global_myname(c->opt_requester_name);
+ 	}
+ 
+-	if (!c->opt_user_name && getenv("LOGNAME")) {
+-		c->opt_user_name = getenv("LOGNAME");
++	if (!c->opt_user_name) {
++		if(getenv("LOGNAME"))
++			c->opt_user_name = getenv("LOGNAME");
++		else
++			d_fprintf(stderr,
++				_("Environment LOGNAME is not defined."
++			          " Trying anonymous access.\n"));
+ 	}
+ 
+ 	if (!c->opt_workgroup) {
diff --git net/samba36/files/pkg-deinstall.in net/samba36/files/pkg-deinstall.in
index 29ecc2c..fbe0993 100644
--- net/samba36/files/pkg-deinstall.in
+++ net/samba36/files/pkg-deinstall.in
@@ -31,6 +31,7 @@ post-deinstall() {
 	${RMDIR} ${SAMBA_RUNDIR} 2>/dev/null || true
 	${RMDIR} ${SAMBA_LOCKDIR} 2>/dev/null || true
 	${RMDIR} ${SAMBA_PRIVATEDIR} 2>/dev/null || true
+	${ECHO_CMD}
 	${ECHO_CMD} "WARNING: If you will *NOT* use this package anymore, please remove the"
 	${ECHO_CMD} "  following directories manually:"
 	${ECHO_CMD} "  ${SAMBA_PRIVATEDIR}"
diff --git net/samba36/files/samba.in net/samba36/files/samba.in
index c21fa5f..f45b0f3 100644
--- net/samba36/files/samba.in
+++ net/samba36/files/samba.in
@@ -1,6 +1,6 @@
 #!/bin/sh
 #
-# $FreeBSD: ports/net/samba36/files/samba.in,v 1.3 2012/04/13 09:16:00 dougb Exp $
+# $FreeBSD: head/net/samba36/files/samba.in 311317 2013-01-31 15:59:44Z timur $
 #
 
 # PROVIDE: nmbd smbd
@@ -14,9 +14,9 @@
 # to enable this service:
 #
 #samba_enable="YES"
-# or, for fine grain control:
-#nmbd_enable="YES"
-#smbd_enable="YES"
+# You can disable/enable any of the Samba daemons by specifying:
+#nmbd_enable="NO"
+#smbd_enable="NO"
 %%WINBIND%%# You need to enable winbindd separately, by adding:
 %%WINBIND%%#winbindd_enable="YES"
 #
@@ -28,43 +28,6 @@
 
 name="samba"
 rcvar=samba_enable
-
-# Defaults
-samba_enable=${samba_enable:=NO}
-samba_config_default="%%SAMBA_CONFDIR%%/%%SAMBA_CONFIG%%"
-samba_config=${samba_config="${samba_config_default}"}
-command_args=${samba_config:+-s "${samba_config}"}
-# Fetch parameters from configuration file
-testparm_command="%%PREFIX%%/bin/testparm"
-smbcontrol_command="%%PREFIX%%/bin/smbcontrol"
-samba_parm="${testparm_command} -s -v --parameter-name"
-samba_idmap=$(${samba_parm} 'idmap uid' "${samba_config}" 2>/dev/null)
-samba_lockdir=$(${samba_parm} 'lock directory' "${samba_config}" 2>/dev/null)
-# Load configuration
-load_rc_config "${name}"
-# Setup dependent variables
-if [ -n "${rcvar}" ] && checkyesno "${rcvar}"; then
-    nmbd_enable=${nmbd_enable=YES}
-    smbd_enable=${smbd_enable=YES}
-%%WINBIND%%    # Check that winbindd is actually configured
-%%WINBIND%%    if [ -n "${samba_idmap}" ]; then
-%%WINBIND%%	winbindd_enable=${winbindd_enable=YES}
-%%WINBIND%%    fi
-fi
-# XXX: Hack to enable check of the dependent variables
-eval real_${rcvar}="\${${rcvar}:=NO}"	${rcvar}=YES
-%%WINBIND%%# winbindd
-%%WINBIND%%rcvar=winbindd_enable
-%%WINBIND%%winbindd_enable=${winbindd_enable:=NO}
-%%WINBIND%%winbindd_flags=${winbindd_flags=''}
-# nmbd
-rcvar=nmbd_enable
-nmbd_enable=${nmbd_enable:=NO}
-nmbd_flags=${nmbd_flags="-D"}
-# smbd
-rcvar=smbd_enable
-smbd_enable=${smbd_enable:=NO}
-smbd_flags=${smbd_flags="-D"}
 # Custom commands
 extra_commands="reload status"
 start_precmd="samba_start_precmd"
@@ -75,16 +38,19 @@ restart_precmd="samba_checkconfig"
 reload_precmd="samba_checkconfig"
 reload_cmd="samba_reload_cmd"
 rcvar_cmd="samba_rcvar_cmd"
+# Defaults
+samba_config_default="%%SAMBA_CONFDIR%%/%%SAMBA_CONFIG%%"
+testparm_command="%%PREFIX%%/bin/testparm"
+smbcontrol_command="%%PREFIX%%/bin/smbcontrol"
+samba_parm="${testparm_command} -s -v --parameter-name"
+pid_extra=
 #
 samba_daemons="nmbd smbd"
 %%WINBIND%%samba_daemons="${samba_daemons} winbindd"
-# Requirements
-required_files="${samba_config}"
-required_dirs="${samba_lockdir}"
 
 samba_checkconfig() {
     echo -n "Performing sanity check on Samba configuration: "
-    if ${testparm_command} -s ${samba_config:+"${samba_config}"} >/dev/null 2>&1; then
+    if ${testparm_command} -s ${samba_config:+"${samba_config}"} >/dev/null 2>&1; then #"
 	echo "OK"
     else
 	echo "FAILED"
@@ -94,6 +60,10 @@ samba_checkconfig() {
 }
 
 samba_start_precmd() {
+    # Make sure rundir exists
+    if [ ! -d "%%SAMBA_RUNDIR%%" -a ! -e "%%SAMBA_RUNDIR%%" ]; then
+	install -m 0755 -d "%%SAMBA_RUNDIR%%"
+    fi
     # XXX: Never delete winbindd_idmap, winbindd_cache and group_mapping
     if [ -n "${samba_lockdir}" -a -d "${samba_lockdir}" ]; then
 	echo -n "Removing stale Samba tdb files: "
@@ -108,26 +78,36 @@ samba_start_precmd() {
 }
 
 samba_rcvar_cmd() {
-    local rcvar
+    local name rcvar
     rcvar=${name}_enable
-    eval ${rcvar}=\${real_${rcvar}}
     # Prevent recursive calling
     unset "${rc_arg}_cmd" "${rc_arg}_precmd" "${rc_arg}_postcmd"
     # Check master variable
     run_rc_command "${_rc_prefix}${rc_arg}" ${rc_extra_args}
+    # Check dependent variables
+    for name in ${samba_daemons}; do
+	# XXX
+	rcvars=''; v=''
+	rcvar=${name}_enable
+	run_rc_command "${_rc_prefix}${rc_arg}" ${rc_extra_args}
+    done
 }
 
 samba_reload_cmd() {
-    local name rcvar command pidfile
+    local name rcvar command pidfile force_run
     # Prevent recursive calling
     unset "${rc_arg}_cmd" "${rc_arg}_precmd" "${rc_arg}_postcmd"
+    # Ignore rcvar and run command
+    if [ -n "${_rc_prefix}" -a "${_rc_prefix}" = "one" ] || [ -n "${rc_force}" ] || [ -n "${rc_fast}" ]; then
+	force_run=yes
+    fi
     # Apply to all daemons
     for name in ${samba_daemons}; do
-    	rcvar=${name}_enable
+	rcvar=${name}_enable
 	command="%%PREFIX%%/sbin/${name}"
 	pidfile="%%SAMBA_RUNDIR%%/${name}${pid_extra}.pid"
 	# Daemon should be enabled and running
-	if [ -n "${rcvar}" ] && checkyesno "${rcvar}"; then
+	if ( [ -n "${rcvar}" ] && checkyesno "${rcvar}" ) || [ -n "$force_run" ]; then
 	    if [ -n "$(check_pidfile "${pidfile}" "${command}")" ]; then
 		debug "reloading ${name} configuration"
 		echo "Reloading ${name}."
@@ -139,30 +119,80 @@ samba_reload_cmd() {
 }
 
 samba_cmd() {
-    local name rcvar rcvars v command pidfile samba_daemons result
+    local name rcvar rcvars v command pidfile samba_daemons result force_run
     # Prevent recursive calling
     unset "${rc_arg}_cmd" "${rc_arg}_precmd" "${rc_arg}_postcmd"
     # Stop processes in the reverse order
     if [ "${rc_arg}" = "stop" ] ; then
 	samba_daemons=$(reverse_list ${samba_daemons})
     fi
+    # Ignore rcvar and run command
+    if [ -n "${_rc_prefix}" -a "${_rc_prefix}" = "one" ] || [ -n "${rc_force}" ] || [ -n "${rc_fast}" ]; then
+	force_run=yes
+    fi
     # Assume success
     result=0
     # Apply to all daemons
     for name in ${samba_daemons}; do
-	rcvar=${name}_enable
 	# XXX
 	rcvars=''; v=''
+	rcvar=${name}_enable
 	command="%%PREFIX%%/sbin/${name}"
 	pidfile="%%SAMBA_RUNDIR%%/${name}${pid_extra}.pid"
 	# Daemon should be enabled and running
-	if [ -n "${rcvar}" ] && checkyesno "${rcvar}"; then
+	if ( [ -n "${rcvar}" ] && checkyesno "${rcvar}" ) || [ -n "$force_run" ]; then
 	    run_rc_command "${_rc_prefix}${rc_arg}" ${rc_extra_args}
-	    # If any of the of the commands failed, take it as a total result
+	    # If any of the commands failed, take it as a global result
 	    result=$((${result} || $?))
 	fi
     done
     return ${result}
 }
 
+samba_config_init() {
+    local name
+    # Load configuration
+    load_rc_config "${name}"
+    for name in ${samba_daemons}; do
+	load_rc_config "${name}"
+    done
+    # Location of the config
+    samba_config=${samba_config="${samba_config_default}"} #"
+    # XXX: Hack to work around name change of pid file with non-default config
+    if [ -n "${samba_config}" -a "${samba_config}" != "${samba_config_default}" ]; then
+	pid_extra="-$(basename "${samba_config}")"
+    fi
+    # Setup dependent variables
+    if [ -n "${rcvar}" ] && checkyesno "${rcvar}"; then
+	nmbd_enable=${nmbd_enable=YES}
+	smbd_enable=${smbd_enable=YES}
+    fi
+    # Defaults
+    samba_enable=${samba_enable:=NO}
+    nmbd_enable=${nmbd_enable:=NO}
+    smbd_enable=${smbd_enable:=NO}
+%%WINBIND%%    # Winbindd
+%%WINBIND%%    if [ -z "${winbind_enable}" ]; then
+%%WINBIND%%	samba_idmap=$(${samba_parm} 'idmap uid' "${samba_config}" 2>/dev/null)
+%%WINBIND%%	# Check that winbindd is actually configured
+%%WINBIND%%	if [ -n "${samba_idmap}" ]; then
+%%WINBIND%%		winbindd_enable="YES"
+%%WINBIND%%	fi
+%%WINBIND%%	winbindd_enable=${winbindd_enable:=NO}
+%%WINBIND%%    fi
+}
+
+# Load configuration variables
+samba_config_init
+# Where is the lockdir
+samba_lockdir=$(${samba_parm} 'lock directory' "${samba_config}" 2>/dev/null)
+# Common flags
+command_args=${samba_config:+-s "${samba_config}"} #"
+nmbd_flags=${nmbd_flags="-D"}
+smbd_flags=${smbd_flags="-D"}
+%%WINBIND%%winbindd_flags=${winbindd_flags=''}
+# Requirements
+required_files="${samba_config}"
+required_dirs="${samba_lockdir}"
+
 run_rc_command "$1"
