diff -ru devel/py-ipaddr/Makefile.orig devel/py-ipaddr/Makefile
--- devel/py-ipaddr/Makefile.orig	2010-01-14 13:29:02.000000000 -0600
+++ devel/py-ipaddr/Makefile	2010-12-22 17:06:12.000000000 -0600
@@ -6,7 +6,7 @@
 #
 
 PORTNAME=	ipaddr
-PORTVERSION=	2.0.0
+PORTVERSION=	2.1.5
 CATEGORIES=	devel python
 MASTER_SITES=	${MASTER_SITE_GOOGLE_CODE}
 PKGNAMEPREFIX=	${PYTHON_PKGNAMEPREFIX}
diff -ru devel/py-ipaddr/distinfo.orig devel/py-ipaddr/distinfo
--- devel/py-ipaddr/distinfo.orig	2010-01-14 13:29:02.000000000 -0600
+++ devel/py-ipaddr/distinfo	2010-12-22 17:06:20.000000000 -0600
@@ -1,3 +1,2 @@
-MD5 (ipaddr-2.0.0.tar.gz) = 8eb7f2ba2cd503770a4f648abef11c8b
-SHA256 (ipaddr-2.0.0.tar.gz) = 167f535d6d134a8cf8741aec3a108a69323be1c97962e252204a2aa86ecb4247
-SIZE (ipaddr-2.0.0.tar.gz) = 21678
+SHA256 (ipaddr-2.1.5.tar.gz) = e45f968c7af5fd47ead31faacd2b8ecb64ed32351cf7c353084cee1bb03a88da
+SIZE (ipaddr-2.1.5.tar.gz) = 27237
