diff --git a/devel/llvm/Makefile b/devel/llvm/Makefile
index a7591a6..5d7517d 100644
--- a/devel/llvm/Makefile
+++ b/devel/llvm/Makefile
@@ -8,8 +8,11 @@
 PORTNAME=	llvm
 PORTVERSION=	3.2
 CATEGORIES=	devel lang
-MASTER_SITES=	http://llvm.org/releases/${PORTVERSION}/
-DISTNAME=	${PORTNAME}-${PORTVERSION}.src
+MASTER_SITES=	GH
+
+FETCH_CMD=      git clone git://github.com/mono/llvm llvm-3.2 && tar -pczf llvm-3.2.tar.gz llvm-3.2 && rm -rf llvm-3.2 && echo 
+
+DIST_SUBDIR=    ${PORTNAME}
 
 MAINTAINER=	brooks@FreeBSD.org
 COMMENT=	Low Level Virtual Machine
