Index: sysutils/fusefs-ntfs/Makefile
===================================================================
RCS file: /home/ncvs/ports/sysutils/fusefs-ntfs/Makefile,v
retrieving revision 1.43
diff -u -r1.43 Makefile
--- sysutils/fusefs-ntfs/Makefile	23 Sep 2011 22:25:26 -0000	1.43
+++ sysutils/fusefs-ntfs/Makefile	5 Nov 2012 13:31:25 -0000
@@ -1,12 +1,8 @@
-# New ports collection makefile for:	fusefs-ntfs
-# Date created: 			25 July 2006
-# Whom:					Max Khon
-# $FreeBSD: ports/sysutils/fusefs-ntfs/Makefile,v 1.43 2011/09/23 22:25:26 amdmi3 Exp $
-#
+# Created by: Max Khon
+# $FreeBSD: ports/sysutils/fusefs-ntfs/Makefile,v 1.46 2012/10/01 15:01:14 bdrewery Exp $
 
 PORTNAME=	ntfs
-PORTVERSION=	2011.4.12
-PORTREVISION=	1
+PORTVERSION=	2012.1.15
 CATEGORIES=	sysutils
 MASTER_SITES=	http://tuxera.com/opensource/
 PKGNAMEPREFIX=	fusefs-
@@ -16,8 +12,10 @@
 MAINTAINER=	ports@FreeBSD.org
 COMMENT=	Mount NTFS partitions (read/write) and disk images
 
+LICENSE=	GPLv2
+
 BUILD_DEPENDS=	fusefs-libs>=2.7.2:${PORTSDIR}/sysutils/fusefs-libs
-LIB_DEPENDS=	fuse.2:${PORTSDIR}/sysutils/fusefs-libs
+LIB_DEPENDS=	fuse:${PORTSDIR}/sysutils/fusefs-libs
 RUN_DEPENDS=	${LOCALBASE}/modules/fuse.ko:${PORTSDIR}/sysutils/fusefs-kmod
 
 CONFLICTS_BUILD=	bonobo-1.*
@@ -25,13 +23,16 @@
 USE_LDCONFIG=	yes
 USE_AUTOTOOLS=	libtool
 USE_ICONV=	yes
+USE_PKGCONFIG=	yes
 CONFIGURE_ARGS=	--exec-prefix=${PREFIX} --disable-mount-helper \
 		--disable-mtab --with-fuse=external
 CPPFLAGS+=	-I${WRKSRC}/include/ntfs-3g/ -I${LOCALBASE}/include
 LDFLAGS+=	-L${LOCALBASE}/lib
 
-OPTIONS=	LOCK "Lock the device when mounting (avoids access)" on \
-		UBLIO "Enable user space cache for improved speed" on
+OPTIONS_DEFINE=	LOCK UBLIO
+OPTIONS_DEFAULT=LOCK UBLIO
+LOCK_DESC=	Lock the device when mounting (avoids access)
+UBLIO_DESC=	Enable user space cache for improved speed
 
 SUB_FILES=	pkg-message
 DOCSDIR=	${PREFIX}/share/doc/ntfs-3g
@@ -41,15 +42,15 @@
 		ntfscluster.8 ntfscmp.8 ntfscp.8 ntfsfix.8 ntfsinfo.8 \
 		ntfslabel.8 ntfsls.8 ntfsprogs.8 ntfsresize.8 ntfsundelete.8
 
-.include <bsd.port.pre.mk>
+.include <bsd.port.options.mk>
 
-.if defined(WITH_LOCK)
+.if ${PORT_OPTIONS:MLOCK}
 CFLAGS+=	-DUSE_LOCK
 .endif
 
-.if defined(WITH_UBLIO)
+.if ${PORT_OPTIONS:MUBLIO}
 EXTRA_PATCHES+=	${FILESDIR}/extra-patch-ublio
-LIB_DEPENDS+=	ublio.1:${PORTSDIR}/devel/libublio
+LIB_DEPENDS+=	ublio:${PORTSDIR}/devel/libublio
 CFLAGS+=	-DUSE_UBLIO
 .endif
 
@@ -71,4 +72,4 @@
 .endif
 	@${ECHO_CMD}; ${CAT} ${PKGMESSAGE}; ${ECHO_CMD}
 
-.include <bsd.port.post.mk>
+.include <bsd.port.mk>
Index: sysutils/fusefs-ntfs/distinfo
===================================================================
RCS file: /home/ncvs/ports/sysutils/fusefs-ntfs/distinfo,v
retrieving revision 1.26
diff -u -r1.26 distinfo
--- sysutils/fusefs-ntfs/distinfo	10 Jul 2011 16:29:38 -0000	1.26
+++ sysutils/fusefs-ntfs/distinfo	5 Nov 2012 13:31:25 -0000
@@ -1,2 +1,2 @@
-SHA256 (ntfs-3g_ntfsprogs-2011.4.12.tgz) = aa8c747f0bfe819b2387b05e12db8d35e4ac96dcb2432873a0b939e248b2ee05
-SIZE (ntfs-3g_ntfsprogs-2011.4.12.tgz) = 1126704
+SHA256 (ntfs-3g_ntfsprogs-2012.1.15.tgz) = 6f1611c5000de7ca99141a9b853cba2c8dbd86c8e36d5efbe7ba918af773fb25
+SIZE (ntfs-3g_ntfsprogs-2012.1.15.tgz) = 1149907
Index: sysutils/fusefs-ntfs/pkg-descr
===================================================================
RCS file: /home/ncvs/ports/sysutils/fusefs-ntfs/pkg-descr,v
retrieving revision 1.3
diff -u -r1.3 pkg-descr
--- sysutils/fusefs-ntfs/pkg-descr	6 May 2010 22:50:35 -0000	1.3
+++ sysutils/fusefs-ntfs/pkg-descr	5 Nov 2012 13:31:25 -0000
@@ -4,4 +4,4 @@
 functionality is supported, the major exceptions are changing the file
 ownerships and the access rights.
 
-WWW: http://www.tuxera.com/community/
+WWW: http://www.tuxera.com/community/ntfs-3g-download/
Index: sysutils/fusefs-ntfs/pkg-plist
===================================================================
RCS file: /home/ncvs/ports/sysutils/fusefs-ntfs/pkg-plist,v
retrieving revision 1.20
diff -u -r1.20 pkg-plist
--- sysutils/fusefs-ntfs/pkg-plist	10 Jul 2011 16:29:38 -0000	1.20
+++ sysutils/fusefs-ntfs/pkg-plist	5 Nov 2012 13:31:26 -0000
@@ -36,6 +36,7 @@
 include/ntfs-3g/ntfstime.h
 include/ntfs-3g/object_id.h
 include/ntfs-3g/param.h
+include/ntfs-3g/realpath.h
 include/ntfs-3g/reparse.h
 include/ntfs-3g/runlist.h
 include/ntfs-3g/security.h
@@ -47,7 +48,7 @@
 lib/libntfs-3g.a
 lib/libntfs-3g.la
 lib/libntfs-3g.so
-lib/libntfs-3g.so.81
+lib/libntfs-3g.so.83
 libdata/pkgconfig/libntfs-3g.pc
 sbin/mkntfs
 sbin/ntfsclone
Index: sysutils/fusefs-ntfs/files/README.FreeBSD
===================================================================
RCS file: /home/ncvs/ports/sysutils/fusefs-ntfs/files/README.FreeBSD,v
retrieving revision 1.3
diff -u -r1.3 README.FreeBSD
--- sysutils/fusefs-ntfs/files/README.FreeBSD	11 Oct 2010 01:32:01 -0000	1.3
+++ sysutils/fusefs-ntfs/files/README.FreeBSD	5 Nov 2012 13:31:26 -0000
@@ -18,7 +18,7 @@
 precisely FUSE's library with fuse4bsd's kernel module (port of the kernel
 dependent part of FUSE). For more information see:
 
-NTFS-3G site:	http://ntfs-3g.org/
+NTFS-3G site:	http://www.tuxera.com/community/ntfs-3g-download/
 FUSE site:	http://fuse.sourceforge.net/
 fuse4bsd site:	http://fuse4bsd.creo.hu/
 
Index: sysutils/fusefs-ntfs/files/extra-patch-ublio
===================================================================
RCS file: /home/ncvs/ports/sysutils/fusefs-ntfs/files/extra-patch-ublio,v
retrieving revision 1.6
diff -u -r1.6 extra-patch-ublio
--- sysutils/fusefs-ntfs/files/extra-patch-ublio	10 Jul 2011 16:29:38 -0000	1.6
+++ sysutils/fusefs-ntfs/files/extra-patch-ublio	5 Nov 2012 13:31:26 -0000
@@ -1,9 +1,9 @@
---- libntfs-3g/Makefile.in.orig	2011-04-10 20:04:51.000000000 +0200
-+++ libntfs-3g/Makefile.in	2011-04-25 18:54:01.000000000 +0200
-@@ -250,6 +250,7 @@
+--- ./libntfs-3g/Makefile.in.orig	2012-08-29 19:51:35.000000000 -0500
++++ ./libntfs-3g/Makefile.in	2012-08-29 20:00:20.000000000 -0500
+@@ -251,6 +251,7 @@
+ @INSTALL_LIBRARY_FALSE@noinst_LTLIBRARIES = libntfs-3g.la
  libntfs_3g_la_CFLAGS = $(AM_CFLAGS) -I$(top_srcdir)/include/ntfs-3g
  libntfs_3g_la_LDFLAGS = -version-info $(LIBNTFS_3G_VERSION) -no-undefined
- @FUSE_INTERNAL_TRUE@libntfs_3g_la_LIBADD = $(top_builddir)/libfuse-lite/libfuse-lite.la
 +libntfs_3g_la_LIBADD = -lublio
  libntfs_3g_la_SOURCES = acls.c attrib.c attrlist.c bitmap.c bootsect.c \
  	cache.c collate.c compat.c compress.c debug.c device.c dir.c \
Index: sysutils/fusefs-ntfs/files/patch-src-Makefile.in
===================================================================
RCS file: /home/ncvs/ports/sysutils/fusefs-ntfs/files/patch-src-Makefile.in,v
retrieving revision 1.9
diff -u -r1.9 patch-src-Makefile.in
--- sysutils/fusefs-ntfs/files/patch-src-Makefile.in	11 Oct 2010 01:32:01 -0000	1.9
+++ sysutils/fusefs-ntfs/files/patch-src-Makefile.in	5 Nov 2012 13:31:26 -0000
@@ -1,31 +1,5 @@
 --- src/Makefile.in.orig	2010-03-06 13:12:36.000000000 -0300
 +++ src/Makefile.in	2010-06-21 21:59:54.000000000 -0300
-@@ -255,10 +255,10 @@
- ntfs_3g_LDADD = $(FUSE_LIBS) $(top_builddir)/libntfs-3g/libntfs-3g.la
- @REALLYSTATIC_TRUE@ntfs_3g_LDFLAGS = $(AM_LDFLAGS) -all-static
- ntfs_3g_CFLAGS = \
-+	-I$(top_srcdir)/include/ntfs-3g \
- 	$(AM_CFLAGS) 			\
- 	-DFUSE_USE_VERSION=26 		\
--	$(FUSE_CFLAGS) 			\
--	-I$(top_srcdir)/include/ntfs-3g
-+	$(FUSE_CFLAGS)
- 
- ntfs_3g_SOURCES = ntfs-3g.c
- ntfs_3g_probe_LDADD = $(top_builddir)/libntfs-3g/libntfs-3g.la
-@@ -267,9 +267,9 @@
- @REALLYSTATIC_TRUE@ntfs_3g_probe_LDFLAGS = $(AM_LDFLAGS) -all-static
- @REALLYSTATIC_TRUE@ntfs_3g_usermap_LDFLAGS = $(AM_LDFLAGS) -all-static
- @REALLYSTATIC_TRUE@ntfs_3g_secaudit_LDFLAGS = $(AM_LDFLAGS) -all-static
--ntfs_3g_probe_CFLAGS = $(AM_CFLAGS) -I$(top_srcdir)/include/ntfs-3g
--ntfs_3g_usermap_CFLAGS = $(AM_CFLAGS) -I$(top_srcdir)/include/ntfs-3g
--ntfs_3g_secaudit_CFLAGS = $(AM_CFLAGS) -I$(top_srcdir)/include/ntfs-3g
-+ntfs_3g_probe_CFLAGS = -I$(top_srcdir)/include/ntfs-3g $(AM_CFLAGS)
-+ntfs_3g_usermap_CFLAGS = -I$(top_srcdir)/include/ntfs-3g $(AM_CFLAGS)
-+ntfs_3g_secaudit_CFLAGS = -I$(top_srcdir)/include/ntfs-3g $(AM_CFLAGS)
- ntfs_3g_probe_SOURCES = ntfs-3g.probe.c
- ntfs_3g_usermap_SOURCES = usermap.c
- ntfs_3g_secaudit_SOURCES = secaudit.c
 @@ -667,7 +667,7 @@
  
  info-am:
