--- ports-mgmt/pbi-manager/Makefile	2011-12-07 12:36:38.000000000 -0800
+++ ports-mgmt/pbi-manager/Makefile	2012-05-23 11:36:41.044786953 -0700
@@ -2,10 +2,10 @@
 # Date created:				27 May 2010
 # Whom:					Kris Moore <kmoore@FreeBSD.org>
 #
-# $FreeBSD: ports/ports-mgmt/pbi-manager/Makefile,v 1.7 2011/12/07 20:36:38 kmoore Exp $
+# $FreeBSD: ports/ports-mgmt/pbi-manager/Makefile,v 1.8 2012/05/11 21:04:06 kmoore Exp $
 
 PORTNAME=	pbi-manager
-PORTVERSION=	0.9.7
+PORTVERSION=	0.9.8
 CATEGORIES=	ports-mgmt
 MASTER_SITES=	http://www.pcbsd.org/~kris/software/ \
 		ftp://ftp.pcbsd.org/pub/software/
@@ -13,7 +13,9 @@
 MAINTAINER=	kmoore@FreeBSD.org
 COMMENT=	Push Button Installer Manager
 
+.if !defined(WITHOUT_X11)
 RUN_DEPENDS=	xdg-desktop-icon:${PORTSDIR}/devel/xdg-utils
+.endif
 
 USE_BZIP2=	yes
 NO_BUILD=	yes
--- ports-mgmt/pbi-manager/distinfo	2011-12-07 12:36:38.000000000 -0800
+++ ports-mgmt/pbi-manager/distinfo	2012-05-11 14:04:06.000000000 -0700
@@ -1,2 +1,2 @@
-SHA256 (pbi-manager-0.9.7.tar.bz2) = 45905c40596c85f477cb16bcaa0aa2dfd4028b99c489d35ad434661a4a2a91f2
-SIZE (pbi-manager-0.9.7.tar.bz2) = 225203
+SHA256 (pbi-manager-0.9.8.tar.bz2) = 4a48540adbf112c753662dcc3245a744c878d6fce73dc38428e2f2d004781cc6
+SIZE (pbi-manager-0.9.8.tar.bz2) = 226189
--- /dev/null	1969-12-31 16:00:00.000000000 -0800
+++ ports-mgmt/pbi-manager/files/patch-pbi-manager	2012-05-23 11:39:34.914788481 -0700
@@ -0,0 +1,52 @@
+--- pbi-manager	2012-05-23 11:37:39.424788861 -0700
++++ pbi-manager	2012-05-23 11:25:56.984786225 -0700
+@@ -5551,6 +5551,7 @@
+ 		_flags="${_flags} --tmpfs"
+ 	fi
+ 
++        unset PBI_PROGVERSION
+ 	get_pbi_progversion
+ 
+ 	#echo "Starting build of ${_mp} - ${_pv}"
+@@ -6345,7 +6346,7 @@
+ mk_chroot_file() {
+ 
+ 	# Check if on PC-BSD and we can instead fetch fbsd-release.txz
+-	if [ -e "$PCBSD_ETCCONF" ]; then
++	if [ -e "$PCBSD_ETCCONF" -a -z "$FORCE_FBSD_ONLY" ]; then
+ 
+ 	  cd "$PBI_APPDIR"
+ 
+@@ -6412,6 +6413,16 @@
+ 	
+ 		# Figure out which version of FreeBSD to checkout
+ 		case $_osRel in
++			*ALPHA)
++				local _osMajor=${_osRel%%.*}
++				local svnUrl="svn://svn.freebsd.org/base/releng/${_osMajor}.0"
++				echo "Using $svnUrl for sources"
++				;;
++			*BETA)
++				local _osMajor=${_osRel%%.*}
++				local svnUrl="svn://svn.freebsd.org/base/releng/${_osMajor}.0"
++				echo "Using $svnUrl for sources"
++				;;
+ 			*RELEASE)
+ 				local _osMajor=${_osRel%%.*}
+ 				local svnUrl="svn://svn.freebsd.org/base/releng/${_osMajor}.0"
+@@ -6439,6 +6450,7 @@
+                 else
+                         svn co ${svnUrl} ${PBI_BUILDSRC} >>${PBI_BUILDLOG} 2>>${PBI_BUILDLOG}
+                 fi
++		touch ${PBI_BUILDSRC}/Makefile
+ 
+                 if [ "$?" != "0" ] ; then
+                         if [ "${PBI_DELETE_BUILD}" = "1" ] ; then
+@@ -6450,7 +6462,6 @@
+ 	fi # End of subversion checkout
+ 
+ 	echo "Running buildworld / installworld"
+-	touch ${PBI_BUILDSRC}/Makefile
+ 	cd ${PBI_BUILDSRC}
+         make ${PBI_BUILDFLAGS} buildworld >>${PBI_BUILDLOG} 2>>${PBI_BUILDLOG}
+         if [ "$?" != "0" ] ; then
