Index: net/netatalk/Makefile
===================================================================
RCS file: /home/ncvs/ports/net/netatalk/Makefile,v
retrieving revision 1.103
diff -u -r1.103 net/netatalk/Makefile
--- net/netatalk/Makefile	10 Sep 2011 18:40:45 -0000	1.103
+++ net/netatalk/Makefile	15 Nov 2011 01:09:31 -0000
@@ -78,7 +78,7 @@
 .endif
 
 .if defined(WITH_PAM)
-CONFIGURE_ARGS+=	--with-pam
+CONFIGURE_ARGS+=	--with-pam --with-uams-path=${PREFIX}/libexec/${PORTNAME}-uams/
 PLIST_SUB+=		NETATALKPAM=""
 .else
 CONFIGURE_ARGS+=	--without-pam
Index: net/netatalk/pkg-plist
===================================================================
RCS file: /home/ncvs/ports/net/netatalk/pkg-plist,v
retrieving revision 1.33
diff -u -r1.33 net/netatalk/pkg-plist
--- net/netatalk/pkg-plist	7 Aug 2011 17:43:59 -0000	1.33
+++ net/netatalk/pkg-plist	15 Nov 2011 00:40:18 -0000
@@ -1,5 +1,5 @@
 %%ATFUNCS%%bin/ad
-bin/add_netatalk_printer
+%%APPLETALK%%bin/add_netatalk_printer
 bin/adv1tov2
 %%APPLETALK%%bin/aecho
 bin/afpldaptest
@@ -33,7 +33,7 @@
 %%APPLETALK%%bin/pap
 %%APPLETALK%%bin/papstatus
 %%APPLETALK%%bin/psorder
-bin/showppd
+%%APPLETALK%%bin/showppd
 bin/uniconv
 @unexec if cmp -s %D/etc/AppleVolumes.default %D/etc/AppleVolumes.default.dist; then rm -f %D/etc/AppleVolumes.default; fi
 etc/AppleVolumes.default.dist
@@ -56,36 +56,6 @@
 @unexec if cmp -s %D/etc/papd.conf %D/etc/papd.conf.dist; then rm -f %D/etc/papd.conf; fi
 %%APPLETALK%%etc/papd.conf.dist
 %%APPLETALK%%@exec [ ! -f %B/papd.conf ] && cp %B/%f %B/papd.conf
-libexec/netatalk-uams/uams_clrtxt.so
-libexec/netatalk-uams/uams_dhx.so
-libexec/netatalk-uams/uams_dhx2.so
-%%NETATALKPAM%%libexec/netatalk-uams/uams_dhx2_pam.a
-%%NETATALKPAM%%libexec/netatalk-uams/uams_dhx2_pam.la
-%%NETATALKPAM%%libexec/netatalk-uams/uams_dhx2_pam.so
-libexec/netatalk-uams/uams_dhx2_passwd.a
-libexec/netatalk-uams/uams_dhx2_passwd.la
-libexec/netatalk-uams/uams_dhx2_passwd.so
-%%NETATALKPAM%%libexec/netatalk-uams/uams_dhx_pam.a
-%%NETATALKPAM%%libexec/netatalk-uams/uams_dhx_pam.la
-%%NETATALKPAM%%libexec/netatalk-uams/uams_dhx_pam.so
-libexec/netatalk-uams/uams_dhx_passwd.a
-libexec/netatalk-uams/uams_dhx_passwd.la
-libexec/netatalk-uams/uams_dhx_passwd.so
-%%NETATALKKRB5%%libexec/netatalk-uams/uams_gss.a
-%%NETATALKKRB5%%libexec/netatalk-uams/uams_gss.la
-%%NETATALKKRB5%%libexec/netatalk-uams/uams_gss.so
-libexec/netatalk-uams/uams_guest.a
-libexec/netatalk-uams/uams_guest.la
-libexec/netatalk-uams/uams_guest.so
-%%NETATALKPAM%%libexec/netatalk-uams/uams_pam.a
-%%NETATALKPAM%%libexec/netatalk-uams/uams_pam.la
-%%NETATALKPAM%%libexec/netatalk-uams/uams_pam.so
-libexec/netatalk-uams/uams_passwd.a
-libexec/netatalk-uams/uams_passwd.la
-libexec/netatalk-uams/uams_passwd.so
-libexec/netatalk-uams/uams_randnum.a
-libexec/netatalk-uams/uams_randnum.la
-libexec/netatalk-uams/uams_randnum.so
 include/atalk/acl.h
 include/atalk/adouble.h
 include/atalk/aep.h
@@ -130,29 +100,59 @@
 include/netatalk/phase2.h
 lib/libatalk.a
 lib/libatalk.la
-libexec/etc2ps.sh
-libexec/ifmpap
-libexec/ifmpaprev
-libexec/ifpap
-libexec/ifpaprev
-libexec/ifwmpap
-libexec/ifwmpaprev
-libexec/ifwpap
-libexec/ifwpaprev
-libexec/ofmpap
-libexec/ofpap
-libexec/ofwmpap
-libexec/ofwpap
-libexec/psa
+libexec/netatalk-uams/uams_clrtxt.so
+libexec/netatalk-uams/uams_dhx.so
+libexec/netatalk-uams/uams_dhx2.so
+%%NETATALKPAM%%libexec/netatalk-uams/uams_dhx2_pam.a
+%%NETATALKPAM%%libexec/netatalk-uams/uams_dhx2_pam.la
+%%NETATALKPAM%%libexec/netatalk-uams/uams_dhx2_pam.so
+libexec/netatalk-uams/uams_dhx2_passwd.a
+libexec/netatalk-uams/uams_dhx2_passwd.la
+libexec/netatalk-uams/uams_dhx2_passwd.so
+%%NETATALKPAM%%libexec/netatalk-uams/uams_dhx_pam.a
+%%NETATALKPAM%%libexec/netatalk-uams/uams_dhx_pam.la
+%%NETATALKPAM%%libexec/netatalk-uams/uams_dhx_pam.so
+libexec/netatalk-uams/uams_dhx_passwd.a
+libexec/netatalk-uams/uams_dhx_passwd.la
+libexec/netatalk-uams/uams_dhx_passwd.so
+%%NETATALKKRB5%%libexec/netatalk-uams/uams_gss.a
+%%NETATALKKRB5%%libexec/netatalk-uams/uams_gss.la
+%%NETATALKKRB5%%libexec/netatalk-uams/uams_gss.so
+libexec/netatalk-uams/uams_guest.a
+libexec/netatalk-uams/uams_guest.la
+libexec/netatalk-uams/uams_guest.so
+%%NETATALKPAM%%libexec/netatalk-uams/uams_pam.a
+%%NETATALKPAM%%libexec/netatalk-uams/uams_pam.la
+%%NETATALKPAM%%libexec/netatalk-uams/uams_pam.so
+libexec/netatalk-uams/uams_passwd.a
+libexec/netatalk-uams/uams_passwd.la
+libexec/netatalk-uams/uams_passwd.so
+libexec/netatalk-uams/uams_randnum.a
+libexec/netatalk-uams/uams_randnum.la
+libexec/netatalk-uams/uams_randnum.so
+%%APPLETALK%%libexec/etc2ps.sh
+%%APPLETALK%%libexec/ifmpap
+%%APPLETALK%%libexec/ifmpaprev
+%%APPLETALK%%libexec/ifpap
+%%APPLETALK%%libexec/ifpaprev
+%%APPLETALK%%libexec/ifwmpap
+%%APPLETALK%%libexec/ifwmpaprev
+%%APPLETALK%%libexec/ifwpap
+%%APPLETALK%%libexec/ifwpaprev
+%%APPLETALK%%libexec/ofmpap
+%%APPLETALK%%libexec/ofpap
+%%APPLETALK%%libexec/ofwmpap
+%%APPLETALK%%libexec/ofwpap
+%%APPLETALK%%libexec/psa
 %%APPLETALK%%libexec/psf
-libexec/tfmpap
-libexec/tfmpaprev
-libexec/tfpap
-libexec/tfpaprev
-libexec/tfwmpap
-libexec/tfwmpaprev
-libexec/tfwpap
-libexec/tfwpaprev
+%%APPLETALK%%libexec/tfmpap
+%%APPLETALK%%libexec/tfmpaprev
+%%APPLETALK%%libexec/tfpap
+%%APPLETALK%%libexec/tfpaprev
+%%APPLETALK%%libexec/tfwmpap
+%%APPLETALK%%libexec/tfwmpaprev
+%%APPLETALK%%libexec/tfwpap
+%%APPLETALK%%libexec/tfwpaprev
 sbin/afpd
 %%APPLETALK%%sbin/atalkd
 sbin/cnid_dbd
@@ -160,8 +160,7 @@
 %%APPLETALK%%sbin/papd
 %%TIMELORD%%sbin/timelord
 share/aclocal/netatalk.m4
-%%DATADIR%%/pagecount.ps
-@dirrm share/netatalk
+%%APPLETALK%%%%DATADIR%%/pagecount.ps
+%%APPLETALK%%@dirrm share/netatalk
 @dirrm include/netatalk
 @dirrm include/atalk
-@dirrm libexec/netatalk-uams
